---
title: "Batch Correction"
author: "ALSF CCDL - Candace Savonen"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

*Purpose*: This notebook illustrates how to assess batch effects and how to use
`limma` to batch correct. 

The [refine.bio normalization pipeline]( http://docs.refine.bio/en/latest/main_text.html)
completed prior to your download, may be sufficient for removing batch effects
for many datasets. 
It is good practice to assess for batch effects regardless, especially if your 
dataset consists of more than one type of platform. 

In this example, we will use expression data collected from Acute Myeloid 
Leukemia patients. 
This aggregated dataset consists of four original datasets we aggregated by 
species on refine.bio.

Two of these datasets are RNA-seq data ([SRP048759](https://www.refine.bio/experiments/SRP048759/leucegene-aml-sequencing-part-3) and [SRP033266](https://www.refine.bio/experiments/SRP033266/leucegene-aml-sequencing-part-2)) 
and the other two are measured on a hgu133plus2 microarray ([GSE6891](https://www.refine.bio/experiments/GSE6891/acute-myeloid-leukemia-samples-of-samples-60yrs-on-hg-u133-plus-2) 
and [GSE14468](https://www.refine.bio/experiments/GSE14468/gene-expression-profiling-of-cebpa-double-and-single-mutant-and-cebpa-wild-type-aml)). 

## General requirements for batch correction: 

### 1) Adequate sample sizes per batch
For batch correction to work you need adequate sample size (> 10 at least) for 
each batch.
For this example, the smallest dataset included has 143 samples, and the largest 
dataset having 525 samples. 

### 2) No variable confounds between batches
If batches also have other differences such as tissue type, treatments, or any
other biologically meaningful variable, batch correction cannot be completed 
without severely compromising the integrity and meaning of the data. 

### 3) Pre and post evaluations of the batch correction should be done
Batch correction isn't always necessary, nor is it able to completely get rid of
batch effects. 
Because of this, careful evaluation of your data before and after correction is 
needed. 

## 1) Set up
This will install `limma` for you if you haven't installed it before. 
We recommend looking through [`limma`'s guide](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf)
for further information. 

```{r}
if (!("limma" %in% installed.packages())) {
  # Install limma
  BiocManager::install("limma", update = FALSE)
}
```

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Load ggplot2 for graphing
library(ggplot2)
```

```{r}
# Make a results directory if it isn't created yet
if (!dir.exists("results")) {
  dir.create("results")
}
# Make a plots directory if it isn't created yet
if (!dir.exists("plots")) {
  dir.create("plots")
}
```

## 2) Import and set up data
Data downloaded from refine.bio include a metadata tab separated values ("tsv")
file and a data tsv file. 
These files were obtained from [refine.bio](https://www.refine.bio/) by entering
the search term "AML" and then aggregating by `species`.

```{r Import refine.bio data}
# Put name of data file
data.filename <- file.path( "data", "HOMO_SAPIENS.tsv")

# Unzip the file and direct it to the data folder
#unzip(paste0(data.filename, ".zip"), exdir = "data")

# Read in data tsv file
df <- readr::read_tsv(data.filename, progress = FALSE) %>%
  tibble::column_to_rownames('X1')
# Make the gene column the rownames so the gene names are out of the way for calculations
```

Prep metadata to be in proper format for batch info to be used by ComBat

```{r Prep metadata}
# Metadata file name
metadata.filename <- file.path("data", "metadata_HOMO_SAPIENS.tsv")

# Read in data tsv file
metadata <- readr::read_tsv(metadata.filename, progress = FALSE)

# Put samples in the same order in the dataframe
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if the samples are in the same order
all.equal(metadata$refinebio_accession_code, colnames(df))
```

## 3) Set up batch information. 
For this batch correction, we will want both original dataset information as 
well as platform information. 

Set up the platform information. 

```{r}
# Get rid of spaces to streamline things 
metadata$refinebio_platform <- gsub(" ", "", metadata$refinebio_platform)

# Make into a factor
platform <- as.factor(metadata$refinebio_platform)
```

Set up the dataset of origin information.

```{r}
# Make a dataset vector
dataset <- metadata$series_id

# This is hacky and will be removed after we have a 'dataset' metadata column deployed
dataset[grep("SRR16", metadata$refinebio_accession_code)] <- "SRP048759"
dataset[grep("SRR10", metadata$refinebio_accession_code)] <- "SRP033266"

# Make this a factor
dataset <- as.factor(dataset)

# Look at the summary of how many samples we have per dataset
summary(dataset)
```

## 4) Perform Principal Components Analysis

```{r}
# Calculate PCA
pca <- prcomp(t(df))

# Make into dataframe for plotting with ggplot2
pca.df <- data.frame(pca$x[, 1:2])
```

Plot these PC scores and look for how samples of different platforms and 
datasets cluster. 

```{r}
# Create a scatterplot
pc.plot <- ggplot(pca.df, 
                  aes(x = PC1, y = PC2, color = dataset, shape = platform)) +
  geom_point() + 
  ggtitle("Uncorrected") +
  theme_classic()

# Print out the plot here
pc.plot
```

In this example dataset, we notice this "batch effect" is actually mostly a 
difference in platform. 
The group of samples on either the left or right clusters are from multiple 
different datasets, but on the left is RNA-seq data while on the right is all 
hgu133plus2 microarray data. 

Save plot to a png.

```{r}
# Save the plot to a png
ggsave(file.path("plots", "uncorrected_pca_plot.png"))
```

## 5) Peform batch correction using limma
Here we will use the platforms and datasets as our variables to correct by. 

```{r}
# Use removeBatchEffect function from limma
corrected.df <- limma::removeBatchEffect(df, batch =  platform, batch2 = dataset)
```

## 6) Re-evaluate PCA after batch correction
Let's see if batch correction helped with making PC scores less divided by platform. 

```{r}
# Calculate PCA
corrected.pca <- prcomp(t(corrected.df))

# Make into dataframe for plotting with ggplot2
corrected.pca.df <- data.frame(corrected.pca$x[, 1:2])
```

Plot these batch corrected data PC scores.

```{r}
# Create a scatterplot, same as before, but with the corrected data
pc.plot <- ggplot(corrected.pca.df, 
                  aes(x = PC1, y = PC2, color = dataset, shape = platform)) +
  geom_point() + 
  ggtitle("Batch corrected") +
  theme_classic()

# Print out the plot here
pc.plot
```

Although the correction doesn't completely get rid of the batch effects on PC2,
it clearly reduces the effect of the batch effect on PC1. 
This an example of how batch correction can help, but will probably not get rid 
of the problem entirely. 
Depending on what analyses these data were being used for, we may need to consider 
analyzing these datasets separately, or moving forward with only samples of one 
platform type. 

Save plot with batch corrected PCA plot to a png.

```{r}
ggsave(file.path("plots", "batch_corrected_pca_plot.png"))
```

## 7) Save batch corrected data to tsv file

We will save the batch corrected dataframe to a tsv file for hypothetical future
use. 

```{r}
corrected.df %>% as.data.frame() %>% 
  tibble::rownames_to_column("Gene") %>%
  readr::write_tsv(file.path("results", "batch_corrected_exprs.tsv"))
```

Session info: 

```{r Print session info}
# Print session info 
sessionInfo()
```
