---
title: "Clustering Heatmap - RNA-Seq"
author: "CCDL for ALSF"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

*Purpose*: This notebook illustrates one way that you can use harmonized data 
from refine.bio in downstream analyses.

TODO: This introduction section will be revamped when the intro to an example
analysis (#122: https://github.com/AlexsLemonade/refinebio-examples/pull/122)
gets more polished/determined.

## 1) Install libraries
This script uses the R package `DESeq2` for normalization and the R package
`pheatmap` for clustering and creating a heatmap.
<br/> <i> Citation:</i> Love, M.I., Huber, W., Anders, S. Moderated estimation of fold change and dispersion for RNA-Seq data with DESeq2 Genome Biology 15(12):550 (2014)
<br/> <i> Citation:</i> Raivo Kolde (2019). pheatmap: Pretty Heatmaps. R package version 1.0.12. https://CRAN.R-project.org/package=pheatmap

```{r Install and attach the pheatmap and DESeq2 libraries}
if (!("pheatmap" %in% installed.packages())) {
  # Install pheatmap
  install.packages("pheatmap", update = FALSE)
}

if (!("DESeq2" %in% installed.packages())) {
  # Install DESeq2
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the `pheatmap` and `DESeq2` libraries:

```{r}
# Attach the `pheatmap` library
library(pheatmap)

# Attach the `DESeq2` library
library(DESeq2)

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Set the seed so our results are reproducible:
set.seed(12345)
```

Create output folder.

```{r}
# Define the file path to the plots directory
plots_dir <- "plots" # Replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

## 2) Import and set up data
Data downloaded from refine.bio include a metadata [tab separated values ("tsv")]
(https://en.wikipedia.org/wiki/Tab-separated_values) file and a data tsv file.
This chunk of code will read the both tsv files and add them as data.frame
objects to your environment.

The data that we downloaded from refine.bio for this analysis is a dataset, with
a total of 24 samples (obtained from 6 mice with 4 technical replicates for each
mouse), containing RNA-sequencing results for Group 3 and 4 medulloblastoma
orthotopic xenograft mice with and without digoxin treatment.

You can download this dataset by going to the following link and clicking 
`Download Now`: https://www.refine.bio/experiments/SRP150101/transcriptome-analysis-of-group-3-and-4-medulloblastoma-orthotopic-xenograft-mice-with-digoxin-treatment

```{r Import data from .tsv files}
# Read in metadata tsv file
metadata <- readr::read_tsv(file.path(
  "data", # Replace with path to your metadata file
  "metadata_SRP150101.tsv" # Replace with the name of your metadata file
)) %>%
  dplyr::select(refinebio_accession_code, refinebio_title, refinebio_treatment) %>%
  replace(is.na(.), "None") %>%
  dplyr::mutate(
    groups = dplyr::case_when(
    refinebio_treatment == "no treatment" ~ "control",
    TRUE ~ "digoxin"
    # Replace with group names relevant to your data
    ),
    # Create a variable to hold the sample summary information
    # (excluding the lane information)
    sample_info_summary = gsub("(_L).*", "", refinebio_title),
    # Create a variable to store the mouse ids
    mouse_id = stringr::word(refinebio_title, sep = "-|_", 2, 3),
    # Create a variable to store the subgroup information
    subgroup = dplyr::case_when(
      startsWith(mouse_id, "2555MB") ~ "group 3",
      startsWith(mouse_id, "1078MB") ~ "group 4",
      TRUE ~ "unknown"
  ))

# Read in data tsv file
df <- readr::read_tsv(file.path(
  "data", # Replace with path to your data file
  "SRP150101.tsv" # Replace with the name of your data file
)) %>%
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

## 3) Choose genes of interest
Although you may want to create a heatmap including all of the genes in the set,
alternatively, the heatmap could be created using only genes of interest. 
For this example, we will sort genes by variance, but there are many alternative
criterion by which you may want to sort your genes <i>eg</i> fold change,
t-statistic, membership to a particular gene ontology, so on. 

```{r Choose genes}
# Calculate the variance for each gene
variances <- apply(df, 1, var)

# Determine the upper quartile variance cutoff value
upper_var <- quantile(variances, 0.75)

# Subset the data choosing only genes whose variances are in the upper quartile
df_by_var <- df %>%
  dplyr::filter(variances > upper_var)
```

## How are our technical replicates clustering together?

### 4) Create a DESeqDataset
A `DESeqDataSet` object is required to carry out the normalization step that
you will see later using a function from the `DESeq2` package. This chunk of
code will turn the data.frame (or matrix) into a `DESeqDataSet` object.

```{r Create a DESeqDataset}
# The `DESeqDataSetFromMatrix` function needs the values to be converted to
# integers
## TODO: Link to explanation about `DESeqDataSet` in the RNA-seq header section
df_by_var <- df_by_var %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, round)

# Create a `DESeqDataSet` object
dds <- DESeqDataSetFromMatrix(
  countData = df_by_var,
  colData = metadata,
  design = ~ refinebio_treatment
)
```

### 5) Prepare the metadata
We are going to format the metadata to be used for annotation on the heatmap.

```{r}
# Let's prepare the metadata for the uncollapsed `DESeqData` set object
metadata_uncollapsed <- metadata %>%
  dplyr::select(-c("sample_info_summary", "refinebio_title")) %>%
  tibble::column_to_rownames("refinebio_accession_code")
```

### 6) Define a minimum counts cutoff
We are going to do some pre-filtering to keep only rows with 10 or more reads
total.

```{r Define minimum counts cutoff}
# Let's take a look at the distribution of counts first

# Sum up the counts for each row
uncollapsed_count_sums <- data.frame(count = rowSums(counts(dds)))

# Create a density plot using the sums
ggplot2::ggplot(uncollapsed_count_sums, ggplot2::aes(x = count)) +
  ggplot2::geom_density() +
  ggplot2::geom_vline(ggplot2::aes(xintercept = 10), # This line represents the current cutoff of 10
                      color = "blue",
                      linetype = "dashed") +
  ggplot2::geom_vline(ggplot2::aes(xintercept = mean(count)), # This line represents the mean count value
                                   color = "red",
                                   linetype = "dashed")
```

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(dds)) >= 10
dds <- dds[genes_to_keep, ]
```

### 7) Perform DESeq2 normalization
We are going to use a function from the `DESeq2` package to normalize the data. 

```{r Perform DESeq2 normalization}
# Normalize the data in the `DESeqDataSet` object using the
# `normTransform` function from the `DESEq2` R package
## TODO: Add a note/link to section in documentation on why we use
# `normTransform` vs `vst` and `rlog`
ntd_uncollapsed <- normTransform(dds)
```

### 8) Create a heatmap
To further customize the heatmap, see a vignette for a guide at this [link](https://slowkow.com/notes/pheatmap-tutorial/).

```{r Create a heatmap}
# Create and store the heatmap object
uncollapsed_pheatmap <-
  pheatmap(
    assay(ntd_uncollapsed),
    cluster_rows = FALSE, # Replace with `TRUE` to cluster heatmap by rows (genes in this case)
    cluster_cols = FALSE, # Replace with `TRUE` to cluster heatmap by columns (samples in this case)
    annotation_col = metadata_uncollapsed,
    main = "Technical Replicates Heatmap",
    colorRampPalette(c("deepskyblue",
                       "black",
                       "yellow"))(25
                       )
  )
```

"Now that we've seen our replicates are fairly identical... let's take a look at our experimental question". 

## How are our biological replicates clustering together?

### 4) Create a DESeqDataset
A `DESeqDataSet` object is required to carry out the normalization step that
you will see later using a function from the `DESeq2` package. This chunk of
code will collapse the `DESeqDataSet` object that we created above to account
for technical replicates.

```{r}
# Collapse the replicates using the `collapseReplicates` function from the
# `DESEq2` package on the `DESeqDataSet` object we created in
# "Create a DESeqDataset" from the previous section -- this function combines
# the counts from the technical replicates in our dataset 
dds_collapsed <- collapseReplicates(dds, dds$sample_info_summary)
```

### 5) Prepare the metadata

```{r}
# Now we need to collapse the metadata to match the collapsed `DESeqDataSet`
# object
metadata_collapsed <- metadata %>%
  dplyr::select(-c("refinebio_title", "refinebio_accession_code")) %>%
  dplyr::distinct() %>%
  # Now assign the collapsed column of identifiers as rownmaes
  tibble::column_to_rownames("sample_info_summary")
```

### 6) Define a minimum counts cutoff

```{r}
# Let's take a look at the distribution of counts for the collapsed data

# Sum up the counts for each row
collapsed_count_sums <- data.frame(count = rowSums(counts(dds)))

# Create a density plot using the sums
ggplot2::ggplot(collapsed_count_sums, ggplot2::aes(x = count)) +
  ggplot2::geom_density() +
  ggplot2::geom_vline(ggplot2::aes(xintercept = 10), # This line represents the current cutoff of 10
                      color = "blue",
                      linetype = "dashed") +
  ggplot2::geom_vline(ggplot2::aes(xintercept = mean(count)), # This line represents the mean count value
                                   color = "red",
                                   linetype = "dashed")
```

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(dds_collapsed)) >= 10
dds_collapsed <- dds_collapsed[genes_to_keep, ]
```

### 7) Perform DESeq2 normalization

```{r}
# Now let's normalize the data for the collapsed counts using the
# `normTransform` function from the `DESeq2` package that we used in the
# previous section
ntd_collapsed <- normTransform(dds_collapsed)
```

### 8) Create a heatmap

```{r}
# Now let's try creating a heatmap with the collapsed replicate data
collapsed_pheatmap <-
  pheatmap(
    assay(ntd_collapsed),
    cluster_rows = FALSE,
    cluster_cols = FALSE,
    annotation_col = metadata_collapsed,
    main = "Biological Replicates Heatmap",
    colorRampPalette(c("deepskyblue",
                       "black",
                       "yellow"))(25)
  )
```

In the heatmap above, it looks like not only do our technical replicates cluster together, but our biological replicates also cluster together (within the same technical environment).

### 9) Save heatmap as a png
You can easily switch this to save to a jpeg or tiff by changing the function 
and file name within the function to the respective file suffix.

```{r Save heatmap as a png}
# Open a png file
png(file.path(
  plots_dir,
  "HeatmapSRP150101.png" # Replace file name with a relevant output plot name
))

# Print your heatmap
collapsed_pheatmap

# Close the png file:
dev.off()
```

Print session info:

```{r Print session info}
# Print session info
sessionInfo()
```
