---
title: "Clustering Heatmap - RNA-Seq"
author: "CCDL for ALSF"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

*Purpose*: This notebook illustrates one way that you can use harmonized data 
from refine.bio in downstream analyses.

TODO: This introduction section will be revamped when the intro to an example
analysis (#122: https://github.com/AlexsLemonade/refinebio-examples/pull/122)
gets more polished/determined.

## 1) Install libraries
This script uses the R package `DESeq2` for normalization and the R package
`pheatmap` for clustering and creating a heatmap.
<br/> <i> Citation:</i> Love, M.I., Huber, W., Anders, S. Moderated estimation of fold change and dispersion for RNA-Seq data with DESeq2 Genome Biology 15(12):550 (2014)
<br/> <i> Citation:</i> Raivo Kolde (2019). pheatmap: Pretty Heatmaps. R package version 1.0.12. https://CRAN.R-project.org/package=pheatmap

```{r Install and attach the pheatmap and DESeq2 libraries}
if (!("pheatmap" %in% installed.packages())) {
  # Install pheatmap
  install.packages("pheatmap", update = FALSE)
}

if (!("DESeq2" %in% installed.packages())) {
  # Install DESeq2
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the `pheatmap` and `DESeq2` libraries:

```{r}
# Attach the `pheatmap` library
library(pheatmap)

# Attach the `DESeq2` library
library(DESeq2)

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Set the seed so our results are reproducible:
set.seed(12345)
```

Create output folder.

```{r}
# Define the file path to the plots directory
plots_dir <- "plots" # Replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

## 2) Import and set up data
Data downloaded from refine.bio include a metadata [tab separated values ("tsv")]
(https://en.wikipedia.org/wiki/Tab-separated_values) file and a data tsv file.
This chunk of code will read the both tsv files and add them as data.frame
objects to your environment.

The data that we downloaded from refine.bio for this analysis has 24 samples
(obtained from 6 mice with 4 technical replicates for each mouse), containing
RNA-sequencing results for Group 3 and 4 medulloblastoma orthotopic xenograft
mice with and without digoxin treatment.

TODO: Add link to the `getting started` section, specifically to point out how
one can download the dataset.

Let's import the metadata and count data files.

```{r Import data from .tsv files}
# Read in metadata tsv file
metadata <- readr::read_tsv(file.path(
  "data", # Replace with path to your metadata file
  "metadata_SRP150101.tsv" # Replace with the name of your metadata file
)) %>%
  # We only need the three variables from the metadata for this notebook
  dplyr::select(refinebio_accession_code, refinebio_title, refinebio_treatment)

# Read in data tsv file
df <- readr::read_tsv(file.path(
  "data", # Replace with path to your data file
  "SRP150101.tsv" # Replace with the name of your data file
)) %>%
  # Here we are going to store the gene IDs as rownames so that we can have a
  # numeric matrix to perform calculations on later 
  tibble::column_to_rownames("Gene")
```

Let's take a look at the metadata object that we read into the R environment.

```{r}
head(metadata)
```

We are going to manipulate the metadata and add variables with the information
that we would like to use to annotate the heatmaps.

```{r}
metadata <- metadata %>%
  dplyr::mutate(
    # Create a variable to store the treatment group information
    treatment_group = dplyr::case_when(
      refinebio_treatment == "no treatment" ~ "control",
      TRUE ~ "digoxin"
      # Replace with group names relevant to your data
    ),
    # Create a variable to hold the sample summary information
    # (excluding the lane information)
    sample_info_summary = gsub("(_L).*", "", refinebio_title),
    # Create a variable to store the mouse ids
    mouse_id = stringr::word(refinebio_title, sep = "-|_", 2, 3),
    # Create a variable to store the subgroup information
    subgroup = dplyr::case_when(
      startsWith(mouse_id, "2555MB") ~ "group 3",
      startsWith(mouse_id, "1078MB") ~ "group 4",
      TRUE ~ "unknown"
    )
  )
```

Now let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

## How are our technical replicates clustering together?

Our dataset includes technical replicates because we have replicate
data which were multiple sequencing lanes from the same samples and library (we
were able to determine this information using the metadata and associated
article which can be found at the following link: https://pubmed.ncbi.nlm.nih.gov/30355798/).

Before we look at the biological clustering of our data, we should first
determine how our technical replicates cluster together and how that may affect
our biological interpretation.

### 3a) Create a DESeqDataset
We will be using the `DESeq2` package for normalization, which requires us to
format our data into a `DESeqDataSet` object. We turn the data.frame (or matrix)
into a [`DESeqDataSet` object](TODO: Add link to header section) and specify
which variable labels our experimental groups using the `design` argument. 

```{r Create a DESeqDataset for all replicates}
# The `DESeqDataSetFromMatrix` function needs the values to be converted to
# integers
df <- df %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, round)

# Create a `DESeqDataSet` object
dds_all_reps <- DESeqDataSetFromMatrix(
  countData = df, # This is the data.frame with the counts values for all replicates in our dataset
  colData = metadata, # This is the data.frame with the annotation data for the replicates in the counts data.frame
  design = ~ treatment_group + subgroup # Replace with an appropriate design variable for your analysis
)
```

### 4a) Prepare the metadata for all replicates
We are going to format the metadata to be used for annotation on the technical
replicates heatmap.

```{r Prepare the metadata for all replicates object}
# Let's prepare the annotation data.frame for the uncollapsed `DESeqData` set
# object which will be used to create the technical replicates heatmap
annotation_df_all_reps <- metadata %>%
  # We want to select the variables that we want for annotating the technical
  # replicates heatmap
  dplyr::select(refinebio_accession_code,
                mouse_id,
                treatment_group,
                subgroup) %>% 
  # The `pheatmap` function requires that the rownames of our annotation object
  # matches the column names of our `DESeaDataSet` object
  tibble::column_to_rownames("refinebio_accession_code")
```

### 5a) Define a minimum counts cutoff
We want to filter out the genes that have not been expressed or that have
low expression counts because we want to remove any possible noise from our data
before we normalize the data and create our heatmap. We are going to do some
pre-filtering to keep only genes with 10 or more reads total. Note that rows
represent gene data and the columns represent sample data in our dataset.

```{r Plot distribution of counts for all replicates object}
# Let's take a look at the distribution of counts first

# Sum up the counts for each gene -- which means we want the sum of counts for
# each row as the rows represent gene data and the columns represent sample data
# in our dataset
count_sums_all_reps <- data.frame(count = rowSums(counts(dds_all_reps)))

# Create a density plot using the sums
ggplot2::ggplot(count_sums_all_reps, ggplot2::aes(x = count)) +
  ggplot2::geom_density() +
  ggplot2::geom_vline(ggplot2::aes(xintercept = 10), # This line represents the current cutoff of 10
                      color = "blue",
                      linetype = "dashed") +
  ggplot2::geom_vline(ggplot2::aes(xintercept = mean(count)), # This line represents the mean count value
                                   color = "red",
                                   linetype = "dashed")
```

```{r Define a minimum counts cutoff for all replicates object}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep_all_reps <- rowSums(counts(dds_all_reps)) >= 10
dds_all_reps <- dds_all_reps[genes_to_keep_all_reps, ]
```

### 6a) Perform DESeq2 normalization
We are going to use the `rlog` function from the `DESeq2` package to normalize
the data. 
TODO: Add a note/link to section in documentation on why we use `rlog` vs `vst`
and `normTransform`.

```{r Perform DESeq2 normalization on all replicates object}
# Normalize the data in the `DESeqDataSet` object using the
# `rlog` function from the `DESEq2` R package
dds_norm_all_reps <- rlog(dds_all_reps)
```

### 7a) Choose genes of interest
Although you may want to create a heatmap including all of the genes in the set,
alternatively, the heatmap could be created using only genes of interest. 
For this example, we will sort genes by variance, but there are many alternative
criterion by which you may want to sort your genes <i>eg</i> fold change,
t-statistic, membership to a particular gene ontology, so on. 

```{r Choose genes for all replicates object}
# Calculate the variance for each gene
variances_all_reps <- apply(assay(dds_norm_all_reps), 1, var)

# Determine the upper quartile variance cutoff value
upper_var_all_reps <- quantile(variances_all_reps, 0.75)

# Subset the data choosing only genes whose variances are in the upper quartile
df_by_var_all_reps <- data.frame(assay(dds_norm_all_reps)) %>%
  dplyr::filter(variances_all_reps > upper_var_all_reps)
```

### 8a) Create a heatmap
To further customize the heatmap, see a vignette for a guide at this [link](https://slowkow.com/notes/pheatmap-tutorial/).

```{r Create a heatmap for all replicates object}
# Create and store the technical replicates heatmap object
pheatmap_all_reps <-
  pheatmap(
    df_by_var_all_reps,
    cluster_rows = FALSE, # Replace with `TRUE` to cluster heatmap by rows (genes in this case)
    cluster_cols = FALSE, # Replace with `TRUE` to cluster heatmap by columns (samples in this case)
    annotation_col = annotation_df_all_reps,
    main = "Technical Replicates Heatmap",
    colorRampPalette(c("deepskyblue",
                       "black",
                       "yellow"))(25
                       )
  )
```
It appears that are technical replicates are fairly similar as they appear to
cluster in the heatmap above. We know this because the values in the `mouse_id`
annotation appear to cluster together. In order to get a better visualization of
these clusters, you can try re-running the `pheatmap` function in the code chunk
above but with the `cluster_cols` argument changed to be equal to `TRUE`.

### 9a) Save heatmap as a png
You can easily switch this to save to a jpeg or tiff by changing the function 
and file name within the function to the respective file suffix.

```{r Save all reps heatmap as a png}
# Open a png file
png(file.path(
  plots_dir,
  "HeatmapSRP150101_all_reps.png" # Replace file name with a relevant output plot name
))

# Print your heatmap
pheatmap_all_reps

# Close the png file:
dev.off()
```

Now that we've seen our replicates are fairly identical... let's take a look at our experimental question. 

## How are our biological replicates clustering together?

### 3b) Create a DESeqDataset
We will be using the `DESeq2` package for normalization, which requires us to
format our data into a `DESeqDataSet` object. We turn the data.frame (or matrix)
into a [`DESeqDataSet` object](TODO: Add link to header section) and specify
which variable labels our experimental groups using the `design` argument. 


```{r Create a DESeqDataset for collapsed replicates}
# Collapse the replicates using the `collapseReplicates` function from the
# `DESEq2` package on the `DESeqDataSet` object we created in
# "Create a DESeqDataset" from the previous section -- this function combines
# the counts from the technical replicates in our dataset 
dds_collapsed_reps <- collapseReplicates(dds_all_reps,
                                         dds_all_reps$sample_info_summary)
```

### 4b) Prepare the metadata for the collapsed replicates `DESeqDataSet` object
Now, we are going to format the metadata to be used for annotation on the
biological replicates heatmap.

```{r Prepare the metadata for collapsed replicates object}
# Now we need to collapse the metadata to match the collapsed `DESeqDataSet`
# object
annotation_df_collapsed_reps <- metadata %>%
  # We want to select the variables we will need for annotating the biological
  # replicates heatmap
  dplyr::select(sample_info_summary, mouse_id, treatment_group, subgroup) %>%
  # Use `distinct()` to get rid of the duplicates created from removing the
  # variables above as we do not need to represent the technical replicates here
  dplyr::distinct() %>%
  # The `pheatmap` function requires that the rownames of our annotation object
  # matches the column names of our `DESeaDataSet` object
  tibble::column_to_rownames("sample_info_summary")
```

### 5b) Define a minimum counts cutoff
We want to filter out the genes that have not been expressed or that have
low expression counts because we want to remove any possible noise from our data
before we normalize the data and create our heatmap. We are going to do some
pre-filtering to keep only rows with 10 or more reads total. Note that rows
represent gene data and the columns represent sample data in our dataset.

```{r Plot distribution of counts for collapsed replicates object}
# Let's take a look at the distribution of counts for the collapsed data

# Sum up the counts for each gene -- which means we want the sum of counts for
# each row
count_sums_collapsed_reps <-
  data.frame(count = rowSums(counts(dds_collapsed_reps)))

# Create a density plot using the sums
ggplot2::ggplot(count_sums_collapsed_reps, ggplot2::aes(x = count)) +
  ggplot2::geom_density() +
  ggplot2::geom_vline(ggplot2::aes(xintercept = 10), # This line represents the current cutoff of 10
                      color = "blue",
                      linetype = "dashed") +
  ggplot2::geom_vline(ggplot2::aes(xintercept = mean(count)), # This line represents the mean count value
                                   color = "red",
                                   linetype = "dashed")
```

```{r Define a minimum counts cutoff for collapsed replicates object}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep_collapsed_reps <- rowSums(counts(dds_collapsed_reps)) >= 10
dds_collapsed_reps <- dds_collapsed_reps[genes_to_keep_collapsed_reps, ]
```

### 6b) Perform DESeq2 normalization
We are going to use a function from the `DESeq2` package to normalize the data. 

```{r Perform DESeq2 normalization on collapsed replicates object}
# Now let's normalize the data for the collapsed counts using the
# `rlog` function from the `DESeq2` package that we used in the
# previous section
dds_norm_collapsed_reps <- rlog(dds_collapsed_reps)
```

### 7b) Choose genes of interest
Although you may want to create a heatmap including all of the genes in the set,
alternatively, the heatmap could be created using only genes of interest. 
For this example, we will sort genes by variance, but there are many alternative
criterion by which you may want to sort your genes <i>eg</i> fold change,
t-statistic, membership to a particular gene ontology, so on. 

```{r Choose genes for collapsed replicates object}
# Calculate the variance for each gene
variances_collapsed_reps <- apply(assay(dds_norm_collapsed_reps), 1, var)

# Determine the upper quartile variance cutoff value
upper_var_collapsed_reps <- quantile(variances_collapsed_reps, 0.75)

# Subset the data choosing only genes whose variances are in the upper quartile
df_by_var_collapsed_reps <- data.frame(assay(dds_norm_collapsed_reps)) %>%
  dplyr::filter(variances_collapsed_reps > upper_var_collapsed_reps)
```

### 8b) Create a heatmap
To further customize the heatmap, see a vignette for a guide at this [link](https://slowkow.com/notes/pheatmap-tutorial/).

```{r Create a heatmap for collapsed replicates object}
# Before creating a heatmap here, we need to make sure the column names of the
# collapsed replicates object match the row names of the annotation data.frame 
colnames(df_by_var_collapsed_reps) <-
  gsub("[.]", "-", colnames(df_by_var_collapsed_reps))

# Now let's try creating a heatmap with the collapsed biological replicate data
pheatmap_collapsed_reps <-
  pheatmap(
    df_by_var_collapsed_reps,
    cluster_rows = FALSE, # Replace with `TRUE` to cluster heatmap by rows (genes in this case)
    cluster_cols = FALSE, # Replace with `TRUE` to cluster heatmap by columns (samples in this case)
    annotation_col = annotation_df_collapsed_reps,
    main = "Biological Replicates Heatmap",
    colorRampPalette(c("deepskyblue",
                       "black",
                       "yellow"))(25
                       )
  )
```

In the heatmap above, it looks like not only do our technical replicates cluster
together, but our biological replicates also cluster together (within the same
technical environment). We know this because the subgroups cluster together, and
treatment groups cluster together within those subgroups and technical groups.

### 9b) Save heatmap as a png
You can easily switch this to save to a jpeg or tiff by changing the function 
and file name within the function to the respective file suffix.

```{r Save collapsed reps heatmap as a png}
# Open a png file
png(file.path(
  plots_dir,
  "HeatmapSRP150101_collapsed_reps.png" # Replace file name with a relevant output plot name
))

# Print your heatmap
pheatmap_collapsed_reps

# Close the png file:
dev.off()
```

Print session info:

```{r Print session info}
# Print session info
sessionInfo()
```
