---
title: "Differential Expression Analysis for GSE37382"
author: "ALSF CCDL - Candace Savonen"
output:   
  html_notebook:
    toc: true
    toc_float: true
---

*Purpose*: This notebook tests differential expression in probe data from
author-submitted version of GSE37382, a medulloblastoma microarray dataset.

The results from this differential expression analysis is used in the following
notebook to compare how the processing of this microarray data affects
differential expression outcomes.
The data here are medulloblastoma microarray data produced by [Northcott et al, 2012](https://www.ncbi.nlm.nih.gov/pubmed/22832581) and are obtained from Gene
Expression Omnibus (GEO).

## 1) Install libraries
This script uses the bioconductor R package limma to identify differentially
expressed genes.  
The full guide on [limma](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf)
shows examples of limma functions.  
We will also use the library `GEOquery` to obtain the author-processed data.

```{r Install limma}
if (!("limma" %in% installed.packages())) {
  # Install limma
  BiocManager::install("limma", update = FALSE)
}
if (!("GEOquery" %in% installed.packages())) {
  # Install GEOquery
  BiocManager::install("GEOquery", update = FALSE)
}
```

Attach the `limma` library:

```{r}
# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Attach library
library(limma)
```

Set up output directories.

```{r}
# Define the file path to the results directory
results_dir <- "results" # Replace with path to desired output results directory

# Make a results directory if it isn't created yet
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# Define the file path to the plots directory
plots_dir <- "plots" # Replace with path to desired output plots directory

# Make a plots directory if it isn't created yet
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the data directory
data_dir <- "data" # Replace with path to data directory

# Make a data directory if it isn't created yet
if (!dir.exists(data_dir)) {
  dir.create(data_dir)
}
```

## 2) Import and set up author processed data from GEO
For more information on how this data was processed, see its
[GEO entry](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37382)

```{r}
# Store file name for gene matrix
matrix.file.name <- file.path(
  data_dir,
  "GSE37382_series_matrix.txt" # Replace with name of your gene matrix file
)

# Download author processed data file directly to the data directory
geo.data <- GEOquery::getGEO("GSE37382", # Replace with the GEO Accession ID relevant to your txt file -- you can search for a GEO Accession ID using the [NCBI's Gene Expression Omnibus database](https://www.ncbi.nlm.nih.gov/geo/).The GEO ID will start with `GSE`, `GSM`, `GDS`, or `GPL` (more information here: https://www.rdocumentation.org/packages/GEOquery/versions/2.38.4/topics/getGEO)
  destdir = data_dir, GSEMatrix = TRUE
)

# If the unzipped version of the file doesn't exist, run this
if (!file.exists(matrix.file.name)) {
  # Unzip the .gz file with the gene matrix in it
  GEOquery::gunzip(paste0(matrix.file.name, ".gz"))
}
```

Read in the gene matrix.

```{r}
# Read in the author processed gene matrix
author.df <- readr::read_tsv(file.path(
  data_dir,
  "GSE37382_series_matrix.txt" # Replace with name of your gene matrix file
),
comment = "!"
) %>%
  # Store probe ids as the rownames
  tibble::column_to_rownames("ID_REF")
```

## 3) Extract metadata for this dataset

```{r}
# Extract metadata
metadata <- data.frame(geo.data[[1]]@phenoData@data) %>%
  # Get rid of columns that have only NAs
  dplyr::select(-which(apply(is.na(.), 2, all)))

# Double check if metadata and data are in same order
all.equal(metadata$geo_accession, colnames(author.df))

# Print out metadata so we can get an idea of what we have
metadata
```

## 4) Set up design matrix
limma needs a numeric design matrix to signify which are treatment and control
samples.

```{r Set up design matrix}
# Set up group names
subgroup <- as.factor(gsub("Group 3|Group 4", "NonSHH", metadata$subgroup.ch1))

# Create the design matrix
des.mat <- model.matrix(~subgroup)

# Make the column name more informative
colnames(des.mat)[2] <- paste0(levels(subgroup), collapse = "vs")
```

## 5) Apply linear model
After applying our data to linear model, in this example we apply empirical Bayes
smoothing and Benjamini-Hochberg multiple testing correction.

```{r Apply linear model}
# Initial fit
fit <- lmFit(author.df, design = des.mat)

# Make contrasts
fit <- eBayes(fit)
```

Obtain summary statistics:

```{r}
# Apply multiple testing correction and obtain stats
stats <- topTable(fit, number = nrow(author.df)) %>%
  tibble::rownames_to_column("affy_probe_ids")

# Print out stats
stats
```

## 6) Double check results directions with a boxplot
Let's pick a probe from the results to double check our output.

```{r}
stats[32, ]
```

According to the data above, we should see that Affymetrix probe id `8154846` should
be higher in `SHH` group data than in `NonSHH` group data.
Let's extract this data and make a boxplot to check if that is what we see.

```{r}
# Extract a gene's data from the original refine.bio data.frame
test.probe <- author.df %>%
  tibble::rownames_to_column("probe_ids") %>%
  dplyr::filter(probe_ids == stats$affy_probe_ids[32]) %>%
  dplyr::select(-probe_ids) %>%
  t()

# Make a dataframe with the group information
box.plot <- data.frame(test.probe, subgroup) %>%
  # Use ggplot2 to make a boxplot from this info
  ggplot(., aes(x = subgroup, y = test.probe)) +
  geom_boxplot() +
  theme_classic() +
  ggtitle(paste("Probe ID:", stats$affy_probe_ids[32]))

# Print the plot here
box.plot
```

Looks like what we would expect to see.

## 7) Write statistics to output results file

```{r  Write results}
# Save summary statistics to a TSV file
readr::write_tsv(stats, file.path(results_dir, "GSE37382_limma_results.tsv"))
```

Print session info:

```{r}
sessionInfo()
```
