---
title: "Getting Started"
author: "CCDL for ALSF"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [About refine.bio](#about-refinebio)
- [About how this tutorial book is structured](#about-how-this-tutorial-book-is-structured)
- [What you need to install to run the examples](#what-you-need-to-install-to-run-the-examples)
  - [Required software](#required-software)
- [How to get the data for these examples](#how-to-get-the-data-for-these-examples)
- [How to use R Markdown Documents](#how-to-use-r-markdown-documents)
- [An important note about file paths and `.Rmd`s](#an-important-note-about-file-paths-and-rmds)
- [Resources for learning R](#resources-for-learning-r)
- [Additional resources from the CCDL](#additional-resources-from-the-ccdl)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## About refine.bio

[refine.bio](https://www.refine.bio/about) is a collection of ready-to-use transcriptomic data!
Publicly available gene expression data is [uniformly processed](http://docs.refine.bio/en/latest/main_text.html#processing-information) and made available for easy download. 
This tutorial has follow-along examples for use with refine.bio downloads. 

## About how this tutorial book is structured 

This tutorial contains follow-along analysis examples for refine.bio gene expression data.
The analysis examples are organized by technology: ["Microarray"](https://alexslemonade.github.io/refinebio-examples/02-microarray/00-intro-to-microarray.html) or ["RNA-seq"](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html), in addition to an ["Advanced Topics"](https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/00-intro-to-advanced-topics.html) section.
Each analysis is self-contained and provides information with how to obtain the dataset used in the example from refine.bio.
We encourage you to download the [`.Rmd`](http://rmarkdown.rstudio.com) and follow the "getting started" section in the example before diving into our analysis examples.

**Each analysis contains:**   

- A README that introduces you to the analyses, concepts, requirements, and workflows for that module.  
- An R Notebook which consists of:  
    - An R markdown (`.Rmd`) file(s) that you can use in RStudio to run the analysis and contains it's own "getting started" section which describes how to download the example dataset from refine.bio.   
    - An nb.html file that is the resulting output of the `.Rmd` file rendered as an HTML file.

## What you need to install to run the examples 

Our tutorial module requires you to install the following software to be able to run the examples. 
These requirements can be installed by following the instructions at these links in the section below.
We recommend installing `devtools` from CRAN (e.g. running `install.packages("devtools")` in R). 

### Required software:
* [**R**](https://cran.r-project.org/)  [@R-base]  
* [**RStudio**](https://www.rstudio.com/products/RStudio/) an [integrated development environment](https://en.wikipedia.org/wiki/Integrated_development_environment) for working with R and [R Notebooks](https://bookdown.org/yihui/rmarkdown/notebook.html) [@RStudio].    
* [**Bioconductor**](https://bioconductor.org/install/) for installing packages from this package repository [@Huber2015].    
* [**tidyverse**](https://www.tidyverse.org/) - we opt for using tidyverse packages for handling and cleaning the data [@Wickham2019].    
* [**devtools**](https://cran.r-project.org/web/packages/devtools/readme/README.html)
will be required for installing some packages from GitHub [@Wickham2020].  

Each example analysis has additional required packages but will check if they are installed and will install them if they are not.
Depending on your particular configuration, sometimes problems occur (here's a [list of the most common R package installation errors and what they mean](http://mazamascience.com/WorkingWithData/?p=1185)).
Each example module directory will include further instructions for how to follow along with the examples.

## How to get the data for these examples

Each analysis will include a link to the example dataset's page on [refine.bio](https://www.refine.bio/) as well as step-by-step instructions of how to set up your data and folders for your analysis.
As you become more comfortable with the analysis, we encourage you to switch out this example dataset for another that may better suit your interests and scientific questions.
We've placed comments where the code will absolutely need to be changed for a different dataset, but you may find other parts of the analysis you will want to alter to fit your needs.
Each analysis also includes links to resources and documentation that we hope helps you make useful alterations to the analysis. 
You will likely encounter errors and bugs as you make changes; don't let this discourage you, it is all a part of the process.
See this [debugging guide](https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00b-debugging_resources.md) for our list of the most common R errors and how you might be able to address them. 

## How to use R Markdown Documents

We use [R Markdown](https://bookdown.org/yihui/rmarkdown/) throughout this tutorial.
R Markdown documents are helpful for scientific code by allowing you to keep detailed notes, code, and output in one place.

When you execute code within the notebook, the results appear beneath the code. 
Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r print}
print("The output from the code in this chunk will print below!")
```

R Markdown documents also have the added benefit of producing HTML file output that is nicely rendered and easy to read.
Saving one of our R Markdowns (the files that end in `.Rmd`) on your computer will create an HTML file containing the code and output to be saved alongside it (will end in `.nb.html`).

See [this guide using to R Notebooks](https://bookdown.org/yihui/rmarkdown/notebook.html#using-notebooks) for more information about inserting and executing code chunks.

## An important note about file paths and `.Rmd`s

A `current directory` refers to where R will look for files or otherwise operate. 
Directories are the folders of files on your computer; a file path is the series of folders leading to the file you are referring to.
R Markdown documents have the `current directory` always set as wherever the `.Rmd` file itself is saved. 
This means all file paths specified in the `.Rmd` must be specified _relative_ to the location of the `.Rmd`. 

For more practice with setting file paths in `.Rmd` files see these:  

- [This handy course chapter from Baumer and Crouser](https://beanumber.github.io/sds192/lab-import.html#paths)
- [The CCDL's intro to R course material](https://alexslemonade.github.io/training-modules/intro-to-R-tidyverse/01-intro_to_base_R.nb.html)

## Resources for learning R

+ [Swirl, an interactive tutorial](https://swirlstats.com/)  
+ [R for Data Science](https://r4ds.had.co.nz/)  
+ [R Markdown](http://rmarkdown.rstudio.com)  
+ [Tutorial on R, RStudio and R Markdown](https://ismayc.github.io/rbasics-book/)
+ [Handy R cheatsheets](https://www.rstudio.com/resources/cheatsheets/)  
+ [R Notebooks tutorial](https://bookdown.org/yihui/rmarkdown/)  

## Additional resources from the CCDL 

- [Intro to R and Tidyverse course and exercises](https://github.com/AlexsLemonade/training-modules/tree/master/intro-to-R-tidyverse)
- [Tips for debugging code](https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00b-debugging_resources.md)
- [Good scientific coding practices](https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/00c-good-scientific-coding-practices.md)

## References
