---
title: "Introduction to Microarray"
author: "CCDL for ALSF"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Introduction to Microarray technology](#introduction-to-microarray-technology)
  - [Microarray data **strengths**:](#microarray-data-strengths)
  - [Microarray data **limitations**:](#microarray-data-limitations)
- [About quantile normalization](#about-quantile-normalization)
  - [More resources on microarray technology:](#more-resources-on-microarray-technology)
    - [Why doesn't the gene I care about show up in this microarray dataset?](#why-doesnt-the-gene-i-care-about-show-up-in-this-microarray-dataset)
- [References](#references)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Introduction to Microarray technology

Data analyses are generally not "one size fits all"; this is particularly true between RNA-seq vs microarray data.
This tutorial has example analyses [organized by technology](../01-getting-started/getting-started.html#about-how-this-tutorial-book-is-structured) so you can follow examples that are more closely tailored to the nature of the data at hand. 

As with all experimental methods, microarray has strengths and limitations that you should consider in regards to your scientific questions. 

### Microarray data **strengths**:  

- Less expensive than RNA-seq - can afford more replicates and get higher statistical power [@Tarca2006]. 
- Microarray has a faster turn-around than RNA-seq [@intro-microarray-slides].
- It's a simpler experimental procedure than RNA-seq [@intro-microarray-slides].

### Microarray data **limitations**:  

- If a transcript doesn't have a probe designed to it, you won't get it measured; can't be used for transcript discovery [@Mantione2014]. 
- A chip's probe designs are only as up to date as the genome annotation at the time it was designed [@Mantione2014].
- As is true for all techniques that involve nucleotide hybiridization (RNA-seq too); microarray probes come with some biases depending on their nucleotide sequence composition. 

To help address these within-chip probe biases, refine.bio uses [SCAN](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3508193/pdf/nihms401888.pdf) normalization methods [@Piccolo2012]

## About quantile normalization 

Microarray data is generally processed in groups of chips - this can lead to [batch effects](https://en.wikipedia.org/wiki/Batch_effect#:~:text=In%20molecular%20biology%2C%20a%20batch,of%20interest%20in%20an%20experiment). 
To minimize this, all refine.bio microarray data downloads come [quantile-normalized](https://en.wikipedia.org/wiki/Quantile_normalization). 
See the refine.bio docs for more about the microarray process steps including the [quantile normalization](http://docs.refine.bio/en/latest/main_text.html#quantile-normalization).

### More resources on microarray technology: 

- [Getting started in gene expression microarray analysis](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000543) [@Slonim2009].
- [Microarray and its applications](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3467903/) [@Govindarajan2012].
- [An tutorial review of microarray data analysis](http://www.ub.edu/stat/docencia/bioinformatica/microarrays/ADM/slides/A_Tutorial_Review_of_Microarray_data_Analysis_17-06-08.pdf) [@Sanchez2008].
- [Analysis of microarray experiments of gene expression profiling](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2435252/) [@Tarca2006].
- [Some nice summary slides about microarray data analysis from Emory](http://web1.sph.emory.edu/users/hwu30/teaching/bioc/GE1.pdf) [@intro-microarray-slides].

#### Why doesn't the gene I care about show up in this microarray dataset?

If a gene is not detected in any of the samples in a set through our processing, it is still kept in the `Gene` column, but it will be reported as `0`.
For microarray data, a `0` is not necessarily indicative of biology; your gene of interest may have not been covered by the [probes designed for the chip that was used](TODO: Put link to refine.bio docs FAQ when https://github.com/AlexsLemonade/refinebio-docs/issues/137 is addressed). 

If the gene you are interested in does not have an Ensembl ID according to the [version of the annotation](TODO: Put link to refine.bio docs FAQ when https://github.com/AlexsLemonade/refinebio-docs/issues/137 is addressed) we used, it will not be reported in any refine.bio download.  

## References
