---
title: "Ortholog Mapping - Microarray"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
    df_print: paged
---

# Purpose of this analysis

This notebook demonstrates how you can use data from the HUGO Gene Nomenclature Committee (HGNC) database to perform ortholog mapping for microarray data obtained from refine.bio.
In this notebook, we will use zebrafish data from refine.bio and annotate it with human gene symbols from HGNC.

⬇️ [**Jump to the analysis code**](#analysis) ⬇️

# How to run this example

For general information about our tutorials and the basic software packages you will need, please see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-this-tutorial-is-structured).
We recommend taking a look at our [Resources for Learning R](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#resources-for-learning-r) if you have not written code in R before. 

## Obtain the `.Rmd` file

To run this example yourself, [download the `.Rmd` for this analysis by clicking this link](https://alexslemonade.github.io/refinebio-examples/02-microarray/ortholog-mapping_microarray_01_ensembl.Rmd).

You can open this `.Rmd` file in RStudio and follow the rest of these steps from there. (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files.)
Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

## Set up your analysis folders 

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using `.Rmd`s](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Create the data folder if it doesn't exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Define the file path to the plots directory
plots_dir <- "plots" # Can replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results" # Can replace with path to desired output results directory

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/GSE71270/creb-overexpression-induces-leukemia-in-zebrafish-by-blocking-myeloid-differentiation-process).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-email.png" width=500>  

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [CREB overexpression zebrafish dataset](https://www.refine.bio/experiments/GSE71270/creb-overexpression-induces-leukemia-in-zebrafish-by-blocking-myeloid-differentiation-process).
@Tregnago2016 measured microarray gene expression of ten zebrafish samples, five overexpressing human CREB, as well as five control samples.

## Place the dataset in your new `data/` folder

refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#downloadable-files).  

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `GSE71270` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis `.Rmd` you downloaded  
- A folder called "data" which contains:
  - The `GSE71270` folder which contains:
    - The gene expression  
    - The metadata TSV  
- A folder for `plots` (currently empty)  
- A folder for `results` (currently empty)  
    
Your example analysis folder should now look something like this (except with respective experiment accession id and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "GSE71270")

# Check if the gene expression matrix file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "GSE71270.tsv"))

# Check if the metadata file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "metadata_GSE71270.tsv"))
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

***

<!-- Do not delete this line --> <a name="analysis" style="padding-top:56px;margin-top:-56px;">&nbsp;</a>


# Ortholog Mapping - Microarray

## Install libraries

See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources.

Attach the package we need for this analysis.

```{r}
# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import data from HGNC

The HUGO Gene Nomenclature Committee (HGNC) assigns a unique and ideally meaningful name and symbol to every human gene.
The [HGNC database](https://www.genenames.org/) currently contains over 39000 public records containing approved human gene nomenclature and associated gene information [@Gray2015].

The [HGNC Comparison of Orthology Predictions (HCOP)](https://www.genenames.org/tools/hcop/) is a search tool that combines orthology predictions for a specified human gene, or set of human genes from a variety of sources, including Ensembl Compara, HGNC, and NCBI Gene Orthology.
Thus allowing users to obtain and compare orthology data [@Wright2005].
In general, an orthology prediction where most of the databases concur would be considered the reliable, and we will use this to prioritize mapping in cases where there is more than one possible ortholog for a gene.
HCOP was originally designed to show orthology predictions between human and mouse, but has been expanded to include data from 18 genomes, including zebrafish, which we will use in this notebook [@hcop-help].

First, we need to download the file from the FTP server holding the HGNC data.
Go to this [directory page of the HGNC Comparison of Orthology Predictions (HCOP) files](ftp://ftp.ebi.ac.uk/pub/databases/genenames/hcop/).

This is where the files that reflect the data provided via the [HGNC database](https://www.genenames.org/) are maintained.
The files available for download include files that have been mapped between human and another species using the [`hcop` package](https://github.com/stephenturner/hcop/) [@Turner2018].
Ortholog species files with the '6 Column' output returns the raw assertions, Ensembl gene IDs and Entrez Gene IDs for human and one other species, while the '15 Column' output includes additional information such as the chromosomal location, accession numbers and the databases that support the assertions.

To download a file, click the file name.
For this notebook, you will want to download the file named `human_zebrafish_hcop_fifteen_column.txt.gz`.
If you are using a different dataset, you can replace `zebrafish` in `human_zebrafish_hcop_fifteen_column.txt.gz` with the name of the species you have data for, and click on that file to download.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/07499927a22ed7aa045245e40d692f8e49d7ba52/template/screenshots/download-hgnc-zebrafish-file.png" width=400>

Now copy and paste the `human_zebrafish_hcop_fifteen_column.txt.gz` file into your `data/` folder.

In the next chunk, we will read in the orthology file that was just downloaded.

```{r}
# Define the file path to organism orthology file downloaded from the HGNC database
zebrafish_hgnc_file <- file.path("data", "human_zebrafish_hcop_fifteen_column.txt.gz")
```

```{r}
# Read in the data from HGNC
zebrafish <- readr::read_tsv(zebrafish_hgnc_file) %>%
  set_names(names(.) %>%
    gsub("_gene$", "", .)) # Removing some extra wording in some of the column names
```

Let's take a look at what `zebrafish` looks like.

```{r}
zebrafish
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the data TSV file and add it as an object to your environment.

```{r}
# Read in data TSV file
zebrafish_genes <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "GSE71270.tsv" # Replace with the name of your data file
)) %>%
  # We only want the gene IDs so let's pull the `Gene` column
  dplyr::pull("Gene")
```

## Mapping human gene symbols to zebrafish Ensembl gene IDs

refine.bio data uses Ensembl gene identifiers, which will be in the first column.

```{r}
# Let's take a look at the first 6 items of `zebrafish_genes`
head(zebrafish_genes)
```

Ensembl gene identifiers have different species-specific prefixes. 
In zebrafish, Ensembl gene identifiers begin with `ENSDARG` (in human, `ENSG`, etc.).

Now let's do the mapping!

We're interested in the `human_symbol`, `zebrafish_ensembl`, and `support` columns specifically.
The `support` column contains a list of associated databases that support each assertion.
This column may assist with addressing some of the multi-mappings that we will talk about later.

```{r}
human_zebrafish_key <- zebrafish %>%
  # We'll want to subset zebrafish to only the columns we're interested in
  dplyr::select(zebrafish_ensembl, human_symbol, support)

# Since we ignored the additional columns in `zebrafish`, let's check to see if
# we have any duplicates in our `human_zebrafish_key`
any(duplicated(human_zebrafish_key))
```

We do have duplicates!
We don't want to handle duplicate data, so let's remove those duplicates before moving forward.

```{r}
human_zebrafish_key <- human_zebrafish_key %>%
  # We need to use the `distinct()` function to remove duplicates resulted from
  # ignoring the additional columns in the `zebrafish` object
  dplyr::distinct()
```

Now let's join the mapped data from `human_zebrafish_key` with the gene data in `zebrafish_genes`.

```{r}
# First, we need to convert our vector of zebrafish genes into a data frame
human_zebrafish_mapped_df <- data.frame("Gene" = zebrafish_genes) %>%
  # Now we can join the mapped data
  dplyr::left_join(human_zebrafish_key, by = c("Gene" = "zebrafish_ensembl"))
```

Here's what the new data frame looks like:

```{r}
head(human_zebrafish_mapped_df, n = 25)
```

Let's get a summary of the human gene symbols returned in our mapped data frame, `human_zebrafish_mapped_df`. 

```{r}
# We can use this `count()` function after `group_by()`to get a count of how many
# `zebrafish_ensembl` IDs there are per `human_symbol`s
human_zebrafish_mapped_df %>%
  dplyr::group_by(human_symbol) %>%
  dplyr::count() %>%
  # Sort by highest `n` which would be the human gene symbol with the most
  # mapped zebrafish Ensembl IDs
  dplyr::arrange(desc(n))
```

Looks like we have mapped symbols!

Now, let's get an idea of how many zebrafish Ensembl IDs we have that were not mapped to human gene symbols.

```{r}
sum(is.na(human_zebrafish_mapped_df$human_symbol))
```

We have 463 NAs, which means we have 463 zebrafish Ensembl IDs that were not mapped to human gene symbols.
This is okay because we do not expect everything to map across species.

## Take a look at some multi-mappings

If a zebrafish Ensembl gene ID maps to multiple human symbols, the associated values will get duplicated.
Let's look at the `ENSDARG00000069142` example below.

```{r}
human_zebrafish_mapped_df %>%
  dplyr::filter(Gene == "ENSDARG00000069142")
```

On the other hand, if you were to look at the original data associated to the zebrafish Ensembl IDs, when a human gene symbol maps to multiple zebrafish Ensembl IDs, the values will not get duplicated, but you will have multiple rows associated with that human symbol.
Let's look at the `MATR3` example below.

```{r}
human_zebrafish_mapped_df %>%
  dplyr::filter(human_symbol == "MATR3")
```

## Collapse zebrafish genes mapping to multiple human genes

Remember that if a zebrafish Ensembl gene ID maps to multiple human symbols, the values get duplicated.
We can collapse the multi-mapped values into a list for each Ensembl ID as to not have duplicate values in our data frame.

In the next chunk, we show how we can collapse all the human gene symbols into one column where we store them all for each unique zebrafish Ensembl ID.

```{r}
collapsed_human_symbol_df <- human_zebrafish_mapped_df %>%
  # Group by zebrafish Ensembl IDs
  dplyr::group_by(Gene) %>%
  # Collapse the mapped values in `human_zebrafish_mapped_df` into one column named
  # `all_human_symbols` -- note that we will lose the `support` column in this summarizing step
  dplyr::summarize(all_human_symbols = paste(human_symbol, collapse = ";"))

head(collapsed_human_symbol_df)
```

### Write results to file

Now let's write our list of human gene symbols for each unique zebrafish Ensembl ID results to file.

```{r}
readr::write_tsv(
  collapsed_human_symbol_df,
  file.path(
    results_dir,
    # Replace with a relevant output file name
    "GSE71270_zebrafish_ensembl_to_collapsed_human_gene_symbol.tsv"
  )
)
```

## Collapse human genes mapping to multiple zebrafish genes

Since multiple zebrafish Ensembl gene IDs map to the same human symbol, we can also collapse all Ensembl IDs that each unique human gene symbol mapped to.

A known issue in ortholog mapping is how to handle this data because combining values across zebrafish transcripts using an average may result in the loss of a lot of data and will likely not be representative of the zebrafish biology.

Therefore, we will use the `support` column to decide which mappings to retain.
As we noted earlier, an orthology prediction where most of the databases concur would be considered reliable.
Therefore, where we have multi-mapped zebrafish Ensembl gene IDs, we will take the mappings with more than one database to support the assertion.

```{r}
filtered_zebrafish_ensembl_df <- human_zebrafish_mapped_df %>%
  # Count the number of databases in the support column for each prediction
  dplyr::mutate(n_databases = stringr::str_count(support, ",") + 1) %>%
  # Group by human gene symbols
  dplyr::group_by(human_symbol) %>%
  # Now filter for the rows with more than one database in support for each human gene symbol
  dplyr::filter(n_databases > 1)

head(filtered_zebrafish_ensembl_df)
```

### Write results to file

Now let's write our `filtered_zebrafish_ensembl_df` object, with the _reliable_ zebrafish Ensembl IDs for each unique human gene symbol, to file.

```{r}
readr::write_tsv(
  filtered_zebrafish_ensembl_df,
  file.path(
    results_dir,
    # Replace with a relevant output file name
    "GSE71270_filtered_zebrafish_ensembl_to_human_gene_symbol_list.tsv"
  )
)
```

# Resources for further learning

- [Genenames.org: the HGNC resources in 2015](https://pubmed.ncbi.nlm.nih.gov/25361968/) [@Gray2015]
- [HCOP: the HGNC comparison of orthology predictions search tool](https://pubmed.ncbi.nlm.nih.gov/16284797/) [@Wright2005]
- See more on [HCOP and interpreting the data in HGNC](https://www.genenames.org/help/hcop/)[@hcop-help]
- See our [RNA-seq ortholog mapping notebook](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/ortholog-mapping_rnaseq_01_ensembl.Rmd) as another applicable example, since the steps for this workflow do not change with technology. 

# Session info

At the end of every analysis, before saving your notebook, we recommend printing out your session info. 
This helps make your code more reproducible by recording what versions of software and packages you used to run this. 

```{r}
# Print session info
sessionInfo()
```
