---
title: "Pathway analysis: ssGSEA"
output: 
  html_notebook:
    toc: TRUE
    toc_float: TRUE
author: J. Taroni for ALSF CCDL
date: 2019
---

## Background 

Pathway or gene set analysis methods like Quantitative Set Analysis of Gene 
Expression (QuSAGE) 
([Yaari et al. _NAR_. 2013.](https://doi.org/10.1093/nar/gkt660)) or Gene Set
Enrichment Analysis (GSEA)
([Subramanian et al. _PNAS_. 2005.](https://doi.org/10.1073/pnas.0506580102))
require us to specify group labels.
We may want a better idea of what pathways are up- or down-regulated in 
_individual samples_ if we, for example, suspect that there are subgroups of
patients during exploratory data analysis.
We can use single-sample GSEA (ssGSEA) 
([Barbie et al. _Nature_. 2009.](https://dx.doi.org/10.1038/nature08460)),
which is implemented in the 
[`GSVA` bioconductor package](https://bioconductor.org/packages/release/bioc/html/GSVA.html).
Note that `GSVA` contains _multiple_ gene set enrichment methods and has an 
excellent [vignette](https://bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.pdf).

## Gene sets

We will use KEGG (Kyoto Encyclopedia of Genes and Genomes) Pathways in this
analysis.
We acquired the KEGG `v6.2` pathway set that used human Entrez IDs from the 
[Molecular Signatures Database (MSigDB)](http://software.broadinstitute.org/gsea/msigdb).

In the [`kegg_ortholog_mapping`](./kegg_ortholog_mapping.nb.html) notebook in 
this module, we mapped human Entrez IDs to mouse symbols using the 
[`hcop` package](https://github.com/stephenturner/hcop).
When there was a 1:many mapping between human Entrez IDs and mouse gene symbols,
we selected the mouse gene symbol with the highest number of resources 
supporting the mapping.
This decision might not be suitable for every experiment and may result in
some loss of information.

## Dataset

We're using [`GSE75574`](https://www.refine.bio/experiments/GSE75574/gene-expression-in-mouse-tissues-in-response-to-short-term-calorie-restriction) 
in this notebook. 
This dataset assays the gene expression response to short-term calorie
restriction in multiple tissues from multiple mouse strains.
This is the same dataset we used in one of our QuSAGE examples 
([`qusage_single_dataset`](./qusage_single_dataset.nb.html)), where we only 
looked at white adipose tissue.

## Set up

Package installation and loading.

```{r}
if (!("qusage" %in% installed.packages())) {
  BiocManager::install("qusage", update = FALSE)
}

if (!("org.Mm.eg.db" %in% installed.packages())) {
  BiocManager::install("org.Mm.eg.db", update = FALSE)
}

if (!("GSVA" %in% installed.packages())) {
  BiocManager::install("GSVA", update = FALSE)
}

if (!("matrixStats" %in% installed.packages())) {
  install.packages("matrixStats")
}
```

```{r}
`%>%` <- dplyr::`%>%`
```

```{r}
library(org.Mm.eg.db)
library(GSVA)
```

Create directories to hold plots and results if they do not yet exist.

```{r}
# Define the file path to the plots directory
plots_dir <- "plots" # Replace with path to desired output plots directory

# Make a plots directory if it isn't created yet
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results" # Replace with path to desired output results directory

# Make a results directory if it isn't created yet
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}

# Define the file path to the data directory
data_dir <- "data"
```

## Read in refine.bio data

The gene expression matrix of the dataset we'll be working with is too large 
to be tracked with git without compression, so we need to unzip it if we have
not already.
This compressed folder contains both the expression matrix and the metadata.

```{r}
# if the unzipped folder does not exist
if (!dir.exists(file.path(
  data_dir,
  "GSE75574" # Replace with the name of your file without the .zip extension
))) {
  compressed_file <- file.path(
    data_dir,
    "GSE75574.zip" # Replace with the name of your zipped file
  )
  unzip(compressed_file, exdir = data_dir)
}
```

### Expression data

Because we used this in another notebook, it's possible that you already have
a local copy of the prepared expression matrix.
If a local copy exists, we don't want to go through the steps to reprocess it,
so we can check for this first.

If not, we'll process the data in the following ways:

* Convert the Ensembl gene IDs that are used in refine.bio data to gene symbols.
* If duplicate gene symbols exist, we'll collapse to the mean value for that 
gene symbol.

```{r}
# check for the existence of the file that would have been generated in the
# other notebook
symbol_file <- file.path(
  data_dir,
  "GSE75574", # Replace with name of your unzipped folder
  "GSE75574_symbols.tsv" # Replace with relevant output name
)
if (file.exists(symbol_file)) {
  symbol_exprs_df <- readr::read_tsv(symbol_file, progress = FALSE)
} else {

  # read in the refine.bio expression data
  expression_file <- file.path(
    data_dir,
    "GSE75574", # Replace with name of your unzipped folder
    "GSE75574.tsv" # Replace with name of your expression file
  )
  exprs_df <- readr::read_tsv(expression_file, progress = FALSE)
  # first column is currently named 'Gene' and contains Ensembl gene IDs
  colnames(exprs_df)[1] <- "ENSEMBL"

  # using the default behavior for 1:many mappings here -- select the first one
  symbol_mappings <- mapIds(org.Mm.eg.db,
    keys = exprs_df$ENSEMBL,
    column = "SYMBOL", keytype = "ENSEMBL"
  )

  # mapIds should give us back IDs in the same order as the keys, but we can
  # check to be sure
  if (!(all.equal(names(symbol_mappings), exprs_df$ENSEMBL))) {
    stop("Gene order is not as expected!")
  }

  symbol_exprs_df <- exprs_df %>%
    # add a new column that contains the gene symbols
    # this gets added as the last column
    dplyr::mutate(SYMBOL = symbol_mappings) %>%
    # drop the Ensembl gene IDs
    dplyr::select(-ENSEMBL) %>%
    # reorder such that the gene symbols are in the first column
    dplyr::select(SYMBOL, dplyr::everything()) %>%
    # drop any genes without a gene symbol
    dplyr::filter(!is.na(SYMBOL))

  # if there are any duplicate gene symbols, collapse to the mean value
  if (any(duplicated(symbol_exprs_df$SYMBOL))) {
    symbol_exprs_df <- symbol_exprs_df %>%
      dplyr::group_by(SYMBOL) %>%
      dplyr::summarise_all(mean)
  }

  # write to file
  readr::write_tsv(symbol_exprs_df, symbol_file)
}
```

### Metadata 

```{r}
metadata_file <- file.path(
  data_dir,
  "GSE75574", # Replace with name of your unzipped folder
  "metadata_GSE75574.tsv" # Replace with name of relevant metadata file
)
metadata_df <- readr::read_tsv(metadata_file) %>%
  # drop any metadata columns that are all NAs
  dplyr::select(-which(apply(is.na(.), 2, all)))
```

Retain only the most pertinent columns.

```{r}
metadata_df <- metadata_df %>%
  dplyr::select(
    refinebio_accession_code, refinebio_title,
    refinebio_specimen_part, refinebio_sex, strain
  )
```

We're interested in the are calorie-restricted vs. controls information, which
is in the sample title.
We can extract this and put it in its own column called `condition` as is good
practice.

```{r}
metadata_df <- metadata_df %>%
  dplyr::mutate(condition = dplyr::case_when(
    grepl("control", refinebio_title) ~ "control",
    grepl("calorierestricted", refinebio_title) ~ "calorie-restricted"
  ))
```

## Pathway Analysis

### Read in KEGG pathways

First, we need the sets of genes that represent pathways.
Again, these were prepared in the 
[`kegg_ortholog_mapping`](./kegg_ortholog_mapping.nb.html) notebook (see 
[Gene Sets](#gene-sets) above).

```{r}
kegg_file <- file.path(
  "gene-sets",
  "c2.cp.kegg.v6.2.entrez_mouse_symbol_hcop.gmt"
)
kegg_pathways <- qusage::read.gmt(kegg_file)
```

### Prep expression matrix

```{r}
exprs_mat <- symbol_exprs_df %>%
  tibble::column_to_rownames("SYMBOL") %>%
  # same order as the metdata -- this may be helpful in downstream analyses
  dplyr::select(metadata_df$refinebio_accession_code) %>%
  # qusage takes a matrix
  as.matrix()
```

### ssGSEA

We can specify that we'd like to perform ssGSEA with the `method` argument to
`gsva`.

```{r}
ssgsea_results <- gsva(
  expr = exprs_mat,
  gset.idx.list = kegg_pathways,
  method = "ssgsea",
  verbose = FALSE
)
```

This returns a matrix of ssGSEA enrichment scores where the columns are samples
and the rows are the input gene sets.

```{r}
str(ssgsea_results)
```

#### Identifying "interesting" features

One way to figure out what pathways may be interesting is to find which pathways
have the highest _variance_ in their ssGSEA scores.
We can calculate the row variances with the `matrixStats` package.

```{r}
score_row_var <- matrixStats::rowVars(ssgsea_results)
# for convenience, let's use the pathway names with this vector
names(score_row_var) <- rownames(ssgsea_results)
# look at the 'top' pathways by this metric
head(sort(score_row_var, decreasing = TRUE))
```

Note that many of these pathways seem to have something to do with the immune
system and may very well share a large number of genes.
Let's look at an example pair of these pathways.

```{r}
intersect(
  kegg_pathways$KEGG_ALLOGRAFT_REJECTION,
  kegg_pathways$KEGG_GRAFT_VERSUS_HOST_DISEASE
)
```

These ssGSEA scores should not be treated as if they are independent.

#### Plotting features

What do the patterns across strains and tissues look like?
Let's look at the KEGG Complement and Coagulation Cascades pathway as an 
example.

We'll be using `ggplot2` for plotting, so we'll want our ssGSEA values as a
`data.frame`.

```{r}
comp_coag_df <- data.frame(
  sample_accession = colnames(ssgsea_results),
  ssgsea_score = ssgsea_results["KEGG_COMPLEMENT_AND_COAGULATION_CASCADES", ] # Replace <KEGG_COMPLEMENT_AND_COAGULATION_CASCADES> with desired pathway -- note that this is distinct from `grep` in that you'd have to write the name exactly as it is in the matrix
)
```

We need to add in the strain, tissue, and condition metadata to make an 
informative plot.

```{r}
comp_coag_df <- metadata_df %>%
  dplyr::inner_join(comp_coag_df,
    by = c("refinebio_accession_code" = "sample_accession")
  )
```

```{r fig.height=11, fig.width=8.5}
comp_coag_df %>%
  ggplot(aes(
    x = condition,
    y = ssgsea_score
  )) +
  geom_boxplot() +
  facet_grid(strain ~ refinebio_specimen_part) +
  labs(y = "KEGG Complement and Coagulation Cascades") +
  theme_bw() +
  # x-axis text at a 45 degree angle to increase readability
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    text = element_text(size = 15)
  )
```

```{r}
# saving to file
ggsave(file.path(
  plots_dir,
  "GSE75574_ssgsea_comp_coag_facet.pdf" # Replace with relevant output plot name
),
plot = last_plot() +
  theme(text = element_text(size = 10))
)
```

Looks like there may be some differences between tissues in this pathway, with
some strains showing more pronounced differences in white adipose tissue in the 
calorie-restricted vs. control comparison.
We should always follow up with more formal analysis (e.g., QuSAGE), but 
this series of steps is a good way to explore one's data.

#### Aside: Gene set size influences ssGSEA score values

ssGSEA score values are not necessarily comparable between gene sets of 
different sizes (number of genes).
If you are comparing ssGSEA scores between samples _within a gene set_, this is
not a concern.

To demonstrate the effect of gene set size on scores, we'll perform a short 
experiment with random gene sets.

```{r}
# use all the gene symbols in the dataset as the pool of possible genes
all_genes <- symbol_exprs_df$SYMBOL

# set a seed for reproducibility
set.seed(123)

# GSVA::gsva takes gene sets as a list, so that's how we'll store the random
# gene sets
random_gene_sets <- list()

# testing 5 different gene set sizes
for (pathway_size in c(25, 50, 100, 250, 500)) {
  # generate 10 random sets of pathway_size
  for (path_iter in 1:10) {
    random_pathway_name <- paste("size", pathway_size, path_iter, sep = "_")
    current_gene_set <- base::sample(x = all_genes, size = pathway_size)
    random_gene_sets[[random_pathway_name]] <- current_gene_set
  }
}

# calculate ssGSEA scores for the random gene sets
random_ssgsea_results <- gsva(
  expr = exprs_mat,
  gset.idx.list = random_gene_sets,
  method = "ssgsea",
  verbose = FALSE
)

# plot the results with ggplot2
# first we need to get this in a form that is amenable for plotting, as gsva
# returns a matrix where columns are samples and rows are gene sets
random_long_df <- as.data.frame(random_ssgsea_results) %>%
  # gene set names are rownames
  tibble::rownames_to_column("gene_set") %>%
  # long format
  reshape2::melt(variable.name = "sample", value.name = "ssgsea_score") %>%
  # extract the gene set size from the gene set name
  dplyr::mutate(gene_set_size = stringr::word(gene_set, 2, sep = "_")) %>%
  # we want to plot smallest no. genes -> largest no. genes
  dplyr::mutate(gene_set_size = factor(gene_set_size,
    levels = c(25, 50, 100, 250, 500)
  ))

# violin plot comparing no. genes / gene set size
random_long_df %>%
  ggplot(aes(x = gene_set_size, y = ssgsea_score)) +
  geom_violin() +
  coord_flip() +
  labs(
    title = "Random gene set scores",
    x = "gene set size",
    y = "ssGSEA score"
  ) +
  theme_bw()
```

Note how a smaller gene set results in a larger range of scores.

## Session info

```{r}
sessionInfo()
```
