---
title: "Clustering Data"
author: "ALSF CCDL - Candace Savonen"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---
# Title: "Clustering Heatmap - Microarray"
## Purpose of the analysis: 
Clustering can help you investigate which samples (or genes) are more related to each other.

### What this analysis is good for:

### What this analysis is NOT good for:

## Further resources for learning
 - [Clustering strategies](https://www.statsandr.com/blog/clustering-analysis-k-means-and-hierarchical-clustering-by-hand-and-in-r/)
 - [Complex Heatmap book](https://jokergoo.github.io/ComplexHeatmap-reference/book/)
 - [Complex Heatmap vignette](https://bioconductor.riken.jp/packages/3.4/bioc/vignettes/ComplexHeatmap/inst/doc/s1.introduction.html)
 - [Complex Heatmap publication](https://academic.oup.com/bioinformatics/article/32/18/2847/1743594)

## About dataset we are using for this example
For this example analysis, we will use this [zebrafish microarray dataset](
https://www.refine.bio/experiments/GSE12955/gene-expression-profile-of-zebrafish-kidney-side-population-sp-cells)

## How to obtain the example dataset from refinebio

## How to use this example with a different dataset



*Purpose*: This notebook illustrates one way that you can use harmonized data 
from refine.bio in downstream analyses.

## 1) Install libraries
This script uses the bioconductor R package ComplexHeatmap for clustering and 
creating a heatmap.
<br/> <i> Citation:</i> Gu Z, Eils R, Schlesner M (2016). “Complex heatmaps 
reveal patterns and correlations in multidimensional genomic data.” Bioinformatics.

```{r Install and attach the ComplexHeatmap library}
if (!("ComplexHeatmap" %in% installed.packages())) {
  # Install ComplexHeatmap
  BiocManager::install("ComplexHeatmap", update = FALSE)
}
```

Attach the `ComplexHeatmap` library:

```{r}
# Attach the library
library(ComplexHeatmap)

# Magrittr pipe
`%>%` <- dplyr::`%>%`

# Set the seed so our results are reproducible:
set.seed(12345)
```

Create output folder.

```{r}
# Create the plots folder if it doesn't exist
if (!dir.exists("plots")) {
  dir.create("plots")
}
```

## 2) Import and set up data
Data downloaded from refine.bio include a metadata tab separated values ("tsv")
file and a data tsv file. This chunk of code will read the both tsv files and 
add them as data.frames to your environment.

```{r Import data from .tsv files}
# Read in metadata tsv file
metadata <- readr::read_tsv(file.path("data", "metadata_GSE12955.tsv"))

# Read in data tsv file
df <- readr::read_tsv(file.path("data", "GSE12955.tsv")) %>%
  tibble::column_to_rownames('Gene')
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$geo_accession)

# Check if this is in the same order
all.equal(colnames(df), metadata$geo_accession)
```

## 3) Choose genes of interest
Although you may want to create a heatmap including all of the genes in the set,
alternatively, the heatmap could be created using only genes of interest. 
For this example, we will sort genes by variance, but there are many alternative
criterion by which you may want to sort your genes <i>eg</i> fold change,
t-statistic, membership to a particular gene ontology, so on. 

```{r Choose genes}
# Calculate the variance for each gene
variances <- apply(df, 1, var)

# Determine summary statistics for gene variances
sum.stats.var <- summary(variances)

# Subset the data choosing only genes whose variances are in the upper quartile
df.by.var <- df[which(variances > sum.stats.var[5]), ]
```

## 4) Create a heatmap
To further customize the heatmap, see the vignettes on Bioconductor for a guide 
at this [link](https://bioconductor.org/packages/devel/bioc/vignettes/ComplexHeatmap/inst/doc/s1.introduction.html).

```{r Create a heatmap}
# Reference the ComplexHeatmap guide for further customizing your heatmap
browseVignettes("ComplexHeatmap")

# Create an annotation that labels samples' groups by color
annot <- HeatmapAnnotation(df = data.frame(Groups = rep(c("SP", "MP"), each = 2)),
    col = list(Groups = c("SP" = "green", "MP" = "orange")))

# Create the heatmap object
heatmap <- Heatmap(df.by.var, 
        name = "Gene_Expression",
        show_row_names = FALSE,
        show_row_dend = FALSE,   # Can show the gene/row cluster if this is 
        #changed to TRUE
        column_dend_height = unit(4, "cm"),
        bottom_annotation = annot) # assign the previously made annotation object

# Print out the heatmap
heatmap
```

## 5) Save heatmap as a png
You can easily switch this to save to a jpeg or tiff by changing the function 
and file name within the function to the respective file suffix.

```{r Save heatmap as a png}
# Open a png file
png(file.path("plots", "HeatmapGSE12955.png"))

# Print your heatmap
heatmap

# Close the png file:
dev.off()
```

Print session info:

```{r Print session info}
# Print session info 
sessionInfo()
```
