---
title: "Introduction to RNA-seq"
author: "CCDL for ALSF"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Introduction to RNA-seq technology](#introduction-to-rna-seq-technology)
  - [Further reading about RNA-seq technologies](#further-reading-about-rna-seq-technologies)
- [About DESeq2](#about-deseq2)
  - [What is DESeq2 and why it is handy?](#what-is-deseq2-and-why-it-is-handy)
  - [DESeq2 normalization vs refinebio QN](#deseq2-normalization-vs-refinebio-qn)
  - [Further reading about DESeq2](#further-reading-about-deseq2)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Introduction to RNA-seq technology

Data analyses are generally not "one size fits all"; this is particularly true as we are working with different technologies. 
This tutorial has example analyses [organized by technology](../01-getting-started/getting-started.html#about-how-this-tutorial-book-is-structured) so you can follow examples that are more closely tailored to the nature of the data at hand. 

As with all experimental methods, RNA-seq has strengths and limitations that you should consider in regards to your scientific questions. 

Here's some recommended reading on RNA-seq technology: 
- [StatsQuest: A gentle introduction to RNA-seq](https://www.youtube.com/watch?v=tlf6wYJrwKY)
- [A general background on the wet lab methods of RNA-seq](https://bitesizebio.com/13542/what-everyone-should-know-about-rna-seq/)

#### RNA-seq data **strengths**:

- RNA-seq can collect data on more transcripts (it is less bound to a pre-determined set of probes like microarray is). 
- It's values are considered more dynamic than microarray values which are constrained to the number of probes.
- You can collect information about isoforms, but note that refine.bio's pipeline summarizes to the gene-level so if you are concerned with isoforms, at this time, refine.bio may not be helpful for you. 

#### RNA-seq data **limitations**:

The nature of RNA sequencing steps introduce several different kinds of biases: 
- GC bias - higher GC content sequences are less likely to be observed.
- 3' bias - for most sequencing methods, the 3 prime end of transcripts are more likely to be observed.
- Complexity bias - some sequences are easier to be bound and amplified than others. 
- Library size or sequencing depth - the total number of reads is not always equivalent between samples.
- Gene length - longer genes are more likely to be observed. 

This figure from @Love2016 briefly summarizes these biases. 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/skip-quantile-normalization.png" width=500>

Here's some recommended reading regarding these biases:   
- [Modeling of RNA-seq fragment sequence bias reduces systematic errors in transcript abundance estimation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5143225/) @Love2016.
- [Biases in Illumina transcriptome sequencing caused by random hexamer priming](https://pdfs.semanticscholar.org/9d16/997f5de72d6c606fef3d673db70e5d1d8e1e.pdf?_ga=2.131436679.965169313.1600175795-124991789.1600175795) @Hansen2010.
- [Computation for RNA-seq and ChIP-seq studies](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4121056/) @Pepke2009.
- [StatsQuest: A gentle introduction to RNA-seq](https://www.youtube.com/watch?v=tlf6wYJrwKY) @Starmer2017-rnaseq.

Most normalization methods, including [refine.bio's processing methods](http://docs.refine.bio/en/latest/main_text.html#rna-seq-pipelines), attempt to mitigate these biases, but these biases can never be fully negated.

## About quantile normalization

Refine.bio data is available for you [quantile normalized](https://en.wikipedia.org/wiki/Quantile_normalization), but more often than not, our example modules will recommend using the option for downloading non-quantile normalized data (note that this is RNA-seq specific, and microarray data does not have this download option). 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/e140face75daa6d2c34e30a4755c362e6039a677/template/screenshots/skip-quantile-normalization.png" width=500>

See here for more about the [quantile normalization process in refine.bio](http://docs.refine.bio/en/latest/main_text.html#quantile-normalization)

### Further reading about RNA-seq data and 

Quantify transcripts with Salmon (Patro et al. Nature Methods. 2017.)
Summarize estimates to the gene-level with tximport (Soneson et al. F1000Research. 2015.)
Perform exploratory data analysis with DESeq2 (Love et al. Genome Biology. 2014.)

### Why we use DESeq2
- Plays nicely with tximport which refinebio uses
- It includes a lot of functions that can perform all your basic differential expression needs. 

### DESeq2 normalization 
- Compare and contrast the normalization methods (QN'ed by us vs VST vs rlog vs ???)
- How did we choose which normalization method to use in each tutorial example? 

### Further resources for DESeq2

- [StatsQuest: RPKM, FPKM and TPM, Clearly Explained!!!](https://www.youtube.com/watch?v=TTUrtCY2k-w) @Starmer2015.
- [StatsQuest: StatQuest: DESeq2, part 1, Library Normalization](https://www.youtube.com/watch?v=UFB993xufUU) @Starmer2017-deseq2.
- [DESeq2 vignette: Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) @Love2014
- [Beginner's guide to DESeq2](Bhttps://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf) @Love2014-guide
- [Introduction to DGE - using DESeq2](https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html)

### FAQs about RNA-seq data: 

- Why doesn't the gene I care about show up in this dataset?

There can be a lot of reasons a particular gene might not show up in your refine.bio RNA-seq dataset. 
- The gene was not actually expressed in the cells in the first place.  
- The gene was not expressed at high enough levels to be detectable.
- The gene does not have an Ensembl ID according to the [version of the annotation]() we used. 

- Why do these examples use DESeq2 and not EdgeR or ____

In short, both EdgeR and DESeq2 are good options and we at the CCDL just went with one of our preferences! 
[See this blog that summarizes these - by one of the creators of DESeq2](https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/), he agrees EdgeR is also great. 
But we needed to pick something, and our [refine.bio data is processed by tximport](http://docs.refine.bio/en/latest/main_text.html#processing-information), another package by the same creators of DESeq2, the creators of both these packages designed these packages to all play well together, and [we also really like their documentation and tutorials](#further-resources-for-deseq2). 
But in the end, we are human and have our own preferences. 

If you have strong preferences for EdgeR, you can definitely use your refine.bio data with it, but we currently do not have examples of that. 
In this case, we'd refer you to [EdgeR's users guide](http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf) and wish you the best of luck on your data adventures! 

- What if I care about isoforms?

Unfortunately at this time, all refine.bio data is summarized to the gene-level, and there's no great way to examine isoforms with this data. 
If your research needs to know transcript isoform information, you may need to look elsewhere. 
This [paper discusses some tools for these kinds of questions](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-4002-1) @Zhang2017.

## References
