---
title: "Clustering Heatmap - RNA-seq"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of the analysis:

<!--WHAT DOES THIS ANALYSIS DO AND WHEN WOULD YOU USE IT?-->

This notebook illustrates one way that you can use harmonized RNA-seq data from refine.bio in downstream analyses, specifically in plotting clustered heatmaps.

# How to run this example

For general information about our tutorials, and for the basic software packages you will need; see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-this-tutorial-is-structured).
We recommend taking a look at our [Resources for Learning R](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#resources-for-learning-r) if you have not written code in R before. 

## Obtain the `.Rmd` file of this analysis

To run this example yourself, [download the `.Rmd` for this analysis by clicking this link](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/clustering/clustering_rnaseq_01_heatmap.Rmd). 
You can open this `.Rmd` file in RStudio and follow the rest of these steps from there (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files).
Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

## Set up your analysis folders

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using Rmds](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Define the file path to the results directory
data_dir <- file.path("data", "SRP070849") # Replace with path to desired data directory

# Define the file path to the plots directory
plots_dir <- file.path("plots", "SRP070849") # Can replace with path to desired output plots directory

# Create the data folder if it doesn't exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}
```

In the same place you put this `.Rmd` file, you should now have two new empty folders called `data` and `plots`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/SRP070849).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-email.png" width=500>  

It make take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [acute myeloid leukemia sample dataset](https://www.refine.bio/experiments/SRP070849).

<!--BRIEFLY SUMMARIZE THIS DATASET--> 

The data that we downloaded from refine.bio for this analysis has 19 samples (obtained from 19 mice with acute myeloid leukemia (AML)), containing RNA-sequencing results for types of AML under controlled treatment conditions.

## Place the dataset in your new `data/` folder

Refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#rna-seq-sample-compendium-download-folder).

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`.

Copy and paste both the equivalent `metadata_<experiment_accession_id>.tsv` and `<experiment_accession_id>.tsv` TSV files into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis Rmd you downloaded  
- A folder called "data" which contains:  
  - The gene expression  
  - The metadata TSV  

Your example analysis folder should now look something like this (except with respective experiment accession id and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Check if the gene expression matrix file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "SRP070849.tsv"))

# Check if the metadata file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "metadata_SRP070849.tsv"))
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences.

***

# Clustering Heatmap - RNA-seq

## Install libraries

See our getting started with [package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources. 

In this analysis, we will be using the R package [`DESeq2`](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) [@DESeq2] for normalization and the R package `pheatmap` [@pheatmap] for clustering and creating a heatmap.

```{r}
if (!("pheatmap" %in% installed.packages())) {
  # Install pheatmap
  install.packages("pheatmap", update = FALSE)
}

if (!("DESeq2" %in% installed.packages())) {
  # Install DESeq2
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the `pheatmap` and `DESeq2` libraries:

```{r}
# Attach the `pheatmap` library
library(pheatmap)

# Attach the `DESeq2` library
library(DESeq2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Set the seed so our results are reproducible:
set.seed(12345)
```

## Import and set up data

Data downloaded from refine.bio include a metadata [tab separated values (TSV)](https://en.wikipedia.org/wiki/Tab-separated_values) file and a data TSV file.
This chunk of code will read in both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP070849.tsv" # Replace with the name of your metadata file
))

# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP070849.tsv" # Replace with the name of your data file
)) %>%
  # Here we are going to store the gene IDs as rownames so that we can have a numeric matrix to perform calculations on later
  tibble::column_to_rownames("Gene")
```

Let's take a look at the metadata object that we read into the R environment.

```{r}
head(metadata)
```

Now let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

<!--ALL THE ANALYSIS STEPS HERE -->

Now we are going to use a combination of functions from the `DESeq2` and `pheatmap` packages to look at how are samples and genes are clustering.

## Create a DESeqDataset

We will be using the `DESeq2` package for normalization, which requires us to format our data into a `DESeqDataSet` object.
We turn the data frame (or matrix) into a [`DESeqDataSet` object](TODO: Add link to header section) and specify which variable labels our experimental groups using the [`design` argument](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) [@DESeq2].
In this chunk of code, we will not provide a specific model to the `design` argument because we are not performing a differential expression analysis.

```{r}
# The `DESeqDataSetFromMatrix()` function needs the values to be converted to integers
df <- df %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, round)

# Create a `DESeqDataSet` object
dds <- DESeqDataSetFromMatrix(
  countData = df, # This is the data.frame with the counts values for all replicates in our dataset
  colData = metadata, # This is the data.frame with the annotation data for the replicates in the counts data.frame
  design = ~1 # Here we are not specifying a model -- Replace with an appropriate design variable for your analysis
)
```

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts because we want to remove any possible noise from our data before we normalize the data and create our heatmap.
We are going to do some pre-filtering to keep only genes with 10 or more reads total.
Note that rows represent gene data and the columns represent sample data in our dataset.

```{r}
# Let's take a look at the distribution of counts first

# Sum up the counts for each gene -- which means we want the sum of counts for each row as the rows represent gene data and the columns represent sample data in our dataset
count_sums <- data.frame(count = rowSums(counts(dds)))

# Create a density plot using the sums
ggplot2::ggplot(count_sums, ggplot2::aes(x = count)) +
  ggplot2::geom_density() +
  ggplot2::geom_vline(ggplot2::aes(xintercept = 10), # This line represents the current cutoff of 10
    color = "blue",
    linetype = "solid"
  ) +
  ggplot2::geom_vline(ggplot2::aes(xintercept = mean(count)), # This line represents the mean count value
    color = "red",
    linetype = "dashed"
  )
```

It appears that our arbitrary cutoff of 10 (the solid blue line) is a decent cutoff to make here as it seems to be just about inline with the mean counts value (the red dashed line) and retains the data points that are not in the majority and therefore may hold significance.

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(dds)) >= 10
dds <- dds[genes_to_keep, ]
```

## Perform DESeq2 normalization

We are going to use the `rlog()` function from the `DESeq2` package to normalize the data.
TODO: Add a note/link to section in documentation on why we use `rlog` vs `vst` and `normTransform`.

```{r}
# Normalize the data in the `DESeqDataSet` object using the `rlog()` function from the `DESEq2` R package
dds_norm <- rlog(dds)
```

## Choose genes of interest

Although you may want to create a heatmap including all of the genes in the set, alternatively, the heatmap could be created using only genes of interest.
For this example, we will sort genes by variance, but there are many alternative criterion by which you may want to sort your genes <i>eg</i> fold change, t-statistic, membership to a particular gene ontology, so on. 

```{r}
# Calculate the variance for each gene
variances <- apply(assay(dds_norm), 1, var)

# Determine the upper quartile variance cutoff value
upper_var <- quantile(variances, 0.75)

# Subset the data choosing only genes whose variances are in the upper quartile
df_by_var <- data.frame(assay(dds_norm)) %>%
  dplyr::filter(variances > upper_var)
```

## Create a heatmap

To further customize the heatmap, see a vignette for a guide at this [link](https://slowkow.com/notes/pheatmap-tutorial/) [@make-heatmaps].

```{r}
# Create and store the heatmap object
pheatmap <-
  pheatmap(
    df_by_var,
    cluster_rows = TRUE, # We want to cluster the heatmap by rows (genes in this case)
    cluster_cols = TRUE, # We also want to cluster the heatmap by columns (samples in this case),
    show_rownames = FALSE, # We don't want to show the rownames because there are too many genes for the labels to be clearly seen
    main = "Non-Annotated Heatmap",
    colorRampPalette(c(
      "deepskyblue",
      "black",
      "yellow"
    ))(25
    ),
    scale = "row" # Scale values in the direction of genes (rows)
  )
```

We've created a heatmap but although our genes and samples are clustered, there is not much information that we can gather here because we did not provide the `pheatmap()` function with annotation labels for our samples.

First let's save our clustered heatmap.

### Save heatmap as a png

You can easily switch this to save to a jpeg or tiff by changing the function and file name within the function to the respective file suffix.

```{r}
# Open a png file
png(file.path(
  plots_dir,
  "heatmap_SRP070849_non_annotated.png" # Replace file name with a relevant output plot name
))

# Print your heatmap
pheatmap

# Close the png file:
dev.off()
```

Now, let's add some annotation bars to our heatmap.

## Prepare metadata for annotation

From the accompanying [paper](https://pubmed.ncbi.nlm.nih.gov/28193779/), we know that the mice with `IDH2` mutant AML were treated with vehicle or AG-221 (the first small molecule in vivo inhibitor of IDH2 to enter clinical trials) and the mice with `TET2` mutant AML were treated with vehicle or 5-Azacytidine (Decitabine, hypomethylating agent). [@AML]
We are going to manipulate the metadata and add variables with the information for each sample, from the experimental design briefly described above, that we would like to use to annotate the heatmap.

```{r}
# Let's prepare the annotation data.frame for the uncollapsed `DESeqData` set object which will be used to create the technical replicates heatmap
annotation_df <- metadata %>%
  # Create a variable to store the cancer type information
  dplyr::mutate(
    mutation = dplyr::case_when(
      startsWith(refinebio_title, "TET2") ~ "TET2",
      startsWith(refinebio_title, "IDH2") ~ "IDH2",
      startsWith(refinebio_title, "WT") ~ "WT",
      TRUE ~ "unknown" # If none of the above criteria are true, we mark the `mutation` variable as "unknown"
    )
  ) %>%
  # We want to select the variables that we want for annotating the technical replicates heatmap
  dplyr::select(
    refinebio_accession_code,
    mutation,
    refinebio_treatment
  ) %>%
  # The `pheatmap()` function requires that the row names of our annotation object matches the column names of our `DESeaDataSet` object
  tibble::column_to_rownames("refinebio_accession_code")
```

### Create annotated heatmap

You can create an annotated heatmap by providing our annotation object to the `annotation_col` argument of the `pheatmap()` function.

```{r}
# Create and store the annotated heatmap object
pheatmap_annotated <-
  pheatmap(
    df_by_var,
    cluster_rows = TRUE,
    cluster_cols = TRUE,
    show_rownames = FALSE,
    annotation_col = annotation_df,
    main = "Annotated Heatmap",
    colorRampPalette(c(
      "deepskyblue",
      "black",
      "yellow"
    ))(25
    ),
    scale = "row" # Scale values in the direction of genes (rows)
  )
```

Now that we have annotation bars on our heatmap, we have a better idea of the sample variable groups that appear to cluster together.

Let's save our annotated heatmap.

### Save annotated heatmap as a png

You can easily switch this to save to a JPEG or TIFF by changing the function and file name within the function to the respective file suffix.

```{r}
# Open a PNG file
png(file.path(
  plots_dir,
  "HeatmapSRP070849_annotated.png" # Replace file name with a relevant output plot name
))

# Print your heatmap
pheatmap_annotated

# Close the png file:
dev.off()
```

# Further learning resources about this analysis

- [Make heatmaps in R with pheatmap](https://slowkow.com/notes/pheatmap-tutorial/) [@make-heatmaps]
- To customize heatmaps even further than the functions in the `pheatmap` package allow, see the [ComplexHeatmap Complete Reference Manual](https://jokergoo.github.io/ComplexHeatmap-reference/book/) [@make-complex-heatmaps]

# Print session info:

At the end of every analysis, before saving your notebook, we recommend printing out your session info. 
This helps make your code more reproducible by recording what versions of softwares and packages you used to run this. 

```{r}
# Print session info
sessionInfo()
```
