---
bibliography: /home/rstudio/kitematic/refinebio-examples/references.bib
title: "Differential Expression - RNA-seq"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of the analysis:

<!--WHAT DOES THIS ANALYSIS DO AND WHEN WOULD YOU USE IT?-->

This notebook takes RNA-seq data and metadata from refine.bio and identifies differentially expressed genes based on an experimental groups. 

# How to run this example

For general information about our tutorials, and for the basic software packages you will need; see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-this-tutorial-is-structured).
We recommend taking a look at our [Resources for Learning R](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#resources-for-learning-r) if you have not written code in R before. 

## Obtain the `.Rmd` file of this analysis

To run this example yourself, [download the `.Rmd` for this analysis by clicking this link](https://alexslemonade.github.io/refinebio-examples/ {{RELATIVE PATH TO THIS ANALYSIS FROM REFINEBIO-EXAMPLES}}).

You can open this `.Rmd` file in RStudio and follow the rest of these steps from there (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files).
Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

## Set up your analysis folders 

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using Rmds](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP078441") # Replace with path to desired data directory

# Create the data folder if it doesn't exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Define the file path to the plots directory
plots_dir <- "plots" # Can replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results" # Can replace with path to desired output results directory

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio]({{REFINEBIO DATASET PAGE LINK}}).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-email.png" width=500>  

<!--DELETE THIS SECTION IF YOU DO WANT QUANTILE NORMALIZED DATA--> 

We are going to use non-quantile normalized data for this analysis.
To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/skip-quantile-normalization.png" width=500>
{{DELETE THIS LINE}}

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [acute myeloid leukemia dataset](https://www.refine.bio/experiments/SRP078441/rna-seq-of-primary-patient-aml-samples) [@Micol]

Micol _et._ _al_, performed RNA-seq on primary peripheral blood and bone marrow
samples from AML patients with and without _ASXL1/2_ mutations. 

## Place the dataset in your new `data/` folder

Refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/8f1332ce8c5deb0dd93edf6c84b93bae676d9f3b/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#rna-seq-sample-compendium-download-folder).  

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `<experiment_accession_id>` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis Rmd you downloaded  
- A folder called "data" which contains:
  - The `<experiment_accession_id>` folder which contains:
    - The gene expression  
    - The metadata TSV  

Your example analysis folder should now look something like this (except with respective experiment accession id and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/b78046845763272d0aa65a077b6a3e46977d5342/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Check if the gene expression matrix file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "SRP078441.tsv"))

# Check if the metadata file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "metadata_SRP078441.tsv"))
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

***

# {{NAME OF example}}

## Install libraries

See our getting started with [package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources. 

In this analysis, we will be using DESeq2 [@DESeq2].

```{r}
if (!("DESeq2" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the DESeq2:

```{r}
# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP078441.tsv" # Replace with the name of your metadata file
))

# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP078441.tsv" # Replace with the name of your data file
)) %>%
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

The information we need to make the comparison is in the `refinebio_title` column of the metadata data.frame.

```{r}
head(metadata$refinebio_title)
```

We'll use some tidyverse functionality to extract this information into its own column to make it easier to use.

```{r}
metadata <- metadata %>%
  # the last bit of the title, separated by "-" contains the mutation
  # information that we want to extract
  dplyr::mutate(mutation_status = stringr::word(refinebio_title,
    -1,
    sep = "-"
  )) %>%
  # now let's make this simpler
  dplyr::mutate(mutation_status = dplyr::case_when(
    grepl("ASXL1|ASXL2", mutation_status) ~ "mutation",
    grepl("ASXLwt", mutation_status) ~ "wildtype"
  ))
```

Let's take a look at `metadata_df` to see if this worked.

```{r}
# looking at the first 6 rows of the metadata_df and only at the columns that
# contain the title and the mutation status we extracted from the title
head(dplyr::select(metadata, refinebio_title, mutation_status))
```

### DESeq2

First we need convert the expression data to integers so it can be converted to a DESEq2 object.

```{r}
df <- df %>%
  # Bring the "Gene" column back from the rownames
  tibble::rownames_to_column("Gene") %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, as.integer)
```

Now we need to create `DESeqDataSet` from our expression dataset. 
We use the `mutation_status` of origin in the design formula because that will allow us
to model this variable of interest.

```{r}
ddset <- DESeqDataSetFromMatrix(
  countData = df,
  tidy = TRUE,
  colData = metadata,
  design = ~mutation_status
)
```

### Variance stabilizing transformation

Before visualizing the data, we'll transform it such that it is on a `log2` scale for large counts and library size is taken into account with the `DESeq2` function for variance stabilizing transformation.
See [this section of the `DESeq2` vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization) for more on this topic. 
TODO: Add link to DESeq2 section in RNA-seq intro document. 

```{r vst}
vst_data <- vst(ddset)
```

#### Differential expression analysis

We'll use the wrapper function `DESeq` to do our differential expression analysis.

```{r DESeq}
deseq_object <- DESeq(ddset)
```

Let's take a look at the results table.

```{r deseq_results}
deseq_results <- results(deseq_object)
deseq_results
```

Sort the results. 

```{r arrange-deseq_results}
# this is of class DESeqResults -- we want a data.frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are rownames -- let's make this it's own column for easy
  # display
  tibble::rownames_to_column(var = "Gene") %>%
  dplyr::mutate(threshold = padj < 0.05)

deseq_df %>%
  # let's sort by statistic -- the highest values should be what is up in the
  # MYCN amplified cell lines
  dplyr::arrange(dplyr::desc(stat))
```

Write the `topTable()` output to file.

```{r  Write results}
readr::write_tsv(
  deseq_df,
  file.path(
    results_dir,
    "SRP078441_limma_results.tsv" # Replace with a relevant output file name
  )
)
```

### Volcano plot

```{r}
volcano_plot <- ggplot(deseq_df, aes(x = log2FoldChange, y = -log10(padj), color = threshold)) +
  geom_point() +
  theme_classic() +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  ggtitle("Volcano plot for mutation status") +
  scale_color_manual(values = c("lightgray", "navy")) +
  theme(legend.position = "none")

# Print out plot here
volcano_plot
```

Save the plot to a PNG. 

```{r}
ggsave(
  plot = volcano_plot,
  file.path(plots_dir, "SRP078441_volcano_plot.png")
) # Replace with a plot name relevant to your data
```

Heatmaps are also a pretty common way to show differential expression results. 
You can take your results from this example and make a heatmap following our heatmap module: TODO: link here. 

## Session Info

# Further learning resources about this analysis

- [`DESeq2` vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization)

# Print session info:

At the end of every analysis, before saving your notebook, we recommend printing out your session info. 
This helps make your code more reproducible by recording what versions of softwares and packages you used to run this. 

```{r}
# Print session info
sessionInfo()
```
