---
title: "Differential Expression - RNA-seq"
author: "CCDL for ALSF"
date: "December 2020"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of this analysis

This notebook takes RNA-seq expression data and metadata from refine.bio and identifies differentially expressed genes between two experimental groups. 

Differential expression analysis identifies genes with significantly varying expression among experimental groups by comparing the variation among samples within a group to the variation between groups.
The simplest version of this analysis is comparing two groups where one of those groups is a control group.

Our refine.bio RNA-seq examples use DESeq2 for these analyses because it handles RNA-seq data well and has great documentation.  
Read more about DESeq2 and why we like it on our [Getting Started page](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html#02_About_DESeq2). 

⬇️ [**Jump to the analysis code**](#analysis) ⬇️

# How to run this example

For general information about our tutorials and the basic software packages you will need, please see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-this-tutorial-is-structured).
We recommend taking a look at our [Resources for Learning R](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#resources-for-learning-r) if you have not written code in R before. 

## Obtain the `.Rmd` file

To run this example yourself, [download the `.Rmd` for this analysis by clicking this link](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/differential_expression_rnaseq_01_rnaseq.Rmd).

Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

You can open this `.Rmd` file in RStudio and follow the rest of these steps from there. (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files.)

## Set up your analysis folders 

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using `.Rmd`s](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Create the data folder if it doesn't exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Define the file path to the plots directory
plots_dir <- "plots"

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results"

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data). 

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/SRP123625).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/40e47f4d3f39283effbd9843a457168061be9680/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/40e47f4d3f39283effbd9843a457168061be9680/template/screenshots/download-email.png" width=500>  

We are going to use non-quantile normalized data for this analysis.
To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/40e47f4d3f39283effbd9843a457168061be9680/template/screenshots/skip-quantile-normalization.png" width=500>

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we are using RNA-seq data from an [acute lymphoblastic leukemia (ALL) mouse lymphoid cell model](https://www.refine.bio/experiments/SRP123625) [@Kampen2019]. 
All of the lymphoid mouse cell samples in this experiment have a human RPL10 gene; three with a reference (wild-type) RPL10 gene and three with the R98S mutation. 
We will perform our differential expression using these knock-in and wild-type mice designations. 

## Place the dataset in your new `data/` folder

refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/40e47f4d3f39283effbd9843a457168061be9680/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#downloadable-files).  

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `SRP123625` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis `.Rmd` you downloaded  
- A folder called "data" which contains:
  - The `SRP123625` folder which contains:
    - The gene expression  
    - The metadata TSV  
- A folder for `plots` (currently empty)
- A folder for `results` (currently empty)

Your example analysis folder should now look something like this (except with respective experiment accession ID and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/40e47f4d3f39283effbd9843a457168061be9680/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
These chunks will declare your file paths and double check that your files are in the right place. 

First we will declare our file paths to our data and metadata files, which should be in our data directory.
This is handy to do because if we want to switch the dataset (see next section for more on this) we are using for this analysis, we will only have to change the file path here to get started. 

```{r}
# Define the file path to the data directory
# Replace with the path of the folder the files will be in
data_dir <- file.path("data", "SRP123625")

# Declare the file path to the gene expression matrix file
# inside directory saved as `data_dir`
# Replace with the path to your dataset file
data_file <- file.path(data_dir, "SRP123625.tsv")

# Declare the file path to the metadata file
# inside the directory saved as `data_dir`
# Replace with the path to your metadata file
metadata_file <- file.path(data_dir, "metadata_SRP123625.tsv")
```

Now that our file paths are declared, we can use the `file.exists()` function to check that the files are where we specified above. 

```{r}
# Check if the gene expression matrix file is at the path stored in `data_file`
file.exists(data_file)

# Check if the metadata file is at the file path stored in `metadata_file`
file.exists(metadata_file)
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences.

***

<!-- Do not delete this line --> <a name="analysis" style="padding-top:56px;margin-top:-56px;">&nbsp;</a>


# Differential Expression

## Install libraries

See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources.

In this analysis, we will be using `DESeq2` [@Love2014] for the differential expression testing.
We will also use `EnhancedVolcano` for plotting and `apeglm` for some log fold change estimates  in the results table [@Blighe2020; @Zhu2018]

```{r}
if (!("DESeq2" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("DESeq2", update = FALSE)
}
if (!("EnhancedVolcano" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("EnhancedVolcano", update = FALSE)
}
if (!("apeglm" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("apeglm", update = FALSE)
}
```

Attach the libraries we need for this analysis:

```{r message=FALSE}
# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

The jitter plot we make later on with the `DESeq2::plotCounts()` function involves some randomness. 
As is good practice when our analysis involves randomness, we will set the seed.
<!-- TODO: link to glossary about setting the seed -->

```{r}
set.seed(12345)
```

## Import data and metadata

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

We stored our file paths as objects named `metadata_file` and `data_file` in [this previous step](#check-out-our-file-structure).

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(metadata_file)

# Read in data TSV file
expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
expression_df <- expression_df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(expression_df), metadata$refinebio_accession_code)
```

The information we need to make the comparison is in the `refinebio_title` column of the metadata data.frame.

```{r}
head(metadata$refinebio_title)
```

## Set up metadata

This dataset includes data from mouse lymphoid cells with human RPL10, with and without a `R98S` mutation.
The mutation status is stored along with other information in a single string (this is not very convenient for us). 
We need to extract the mutation status information into its own column to make it easier to use.

```{r}
metadata <- metadata %>%
  # Let's get the RPL10 mutation status from this variable
  dplyr::mutate(mutation_status = dplyr::case_when(
    stringr::str_detect(refinebio_title, "R98S") ~ "R98S",
    stringr::str_detect(refinebio_title, "WT") ~ "reference"
  ))
```

Let's take a look at `metadata` to see if this worked by looking at the `refinebio_title` and `mutation_status` columns. 

```{r}
# Let's take a look at the original metadata column's info
# and our new `mutation_status` column
dplyr::select(metadata, refinebio_title, mutation_status)
```

Before we set up our model in the next step, we want to check if our modeling variable is set correctly. 
We want our "control" to to be set as the first level in the variable we provide as our experimental variable. 
Here we will use the `str()` function to print out a preview of the **str**ucture of our variable

```{r}
# Print out a preview of `mutation_status`
str(metadata$mutation_status)
```

Currently, `mutation_status` is a character, which is not necessarily what we want.
To make sure it is set how we want for the `DESeq` object and subsequent testing, let's change it to a factor so we can explicitly set the levels. 

In the `levels` argument, we will list `reference` first since that is our control group. 

```{r}
# Make mutation_status a factor and set the levels appropriately
metadata <- metadata %>%
  dplyr::mutate(
    # Here we will set up the factor aspect of our new variable.
    mutation_status = factor(mutation_status, levels = c("reference", "R98S"))
  )
```

Note if you don't specify `levels`, the `factor()` function will set levels in alphabetical order -- which sometimes means your control group will not be listed first!

Let's double check if the levels are what we want using the `levels()` function. 

```{r}
levels(metadata$mutation_status)
```

Yes! `reference` is the first level as we want it to be. 
We're all set and ready to move on to making our `DESeq2Dataset` object. 

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts, since these do not have high enough counts to yield reliable differential expression results.
Removing these genes saves on memory usage during the tests. 
We are going to do some pre-filtering to keep only genes with 10 or more reads in total across the samples.

```{r}
# Define a minimum counts cutoff and filter the data to include
# only rows (genes) that have total counts above the cutoff
filtered_expression_df <- expression_df %>%
  dplyr::filter(rowSums(.) >= 10)
```

If you have a bigger dataset, you will probably want to make this cutoff larger. 

## Create a DESeq2Dataset

We will be using the `DESeq2` package for differential expression testing, which requires us to format our data into a `DESeqDataSet` object.
First we need to prep our gene expression data frame so it's in the format that is compatible with the `DESeqDataSetFromMatrix()` function in the next step. 

```{r}
# We are making our data frame into a matrix and rounding the numbers
gene_matrix <- round(as.matrix(filtered_expression_df))
```

Now we need to create `DESeqDataSet` from our expression dataset. 
We use the `mutation_status` variable we created in the design formula because that will allow us
to model the presence/absence of _R98S_ mutation.

```{r}
ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = gene_matrix,
  # Supply the `colData` with our metadata data frame
  colData = metadata,
  # Supply our experimental variable to `design`
  design = ~mutation_status
)
```

## Run differential expression analysis

We'll use the wrapper function `DESeq()` to do our differential expression analysis.
In our `DESeq2` object we designated our `mutation_status` variable as the `model` argument. 
Because of this, the `DESeq` function will use groups defined by `mutation_status` to test for differential expression. 

```{r}
deseq_object <- DESeq(ddset)
```

Let's extract the results table from the `DESeq` object. 

```{r}
deseq_results <- results(deseq_object)
```

Here we will use `lfcShrink()` function to obtain shrunken log fold change estimates based on negative binomial distribution. 
This will add the estimates to your results table. 
Using `lfcShrink()` can help decrease noise and preserve large differences between groups (it requires that `apeglm` package be installed).

```{r}
deseq_results <- lfcShrink(deseq_object, # This is the original DESeq2 object with DESeq() already having been ran
  coef = 2, # This is based on what log fold change coefficient was used in DESeq(), the default is 2.
  res = deseq_results # This needs to be the DESeq2 results table
)
```

Now let's take a peek at what our results table looks like. 

```{r}
head(deseq_results)
```

Note it is not filtered or sorted, so we will use tidyverse to do this before saving our results to a file. 
Sort and filter the results. 

```{r}
# this is of class DESeqResults -- we want a data frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are rownames -- let's make this it's own column for easy
  # display
  tibble::rownames_to_column("Gene") %>%
  dplyr::mutate(threshold = padj < 0.05) %>%
  # sort by statistic -- the highest values should be genes with higher expression
  # RPL10 mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))
```

Let's print out what the top results are. 

```{r}
head(deseq_df)
```

### Check results by plotting one gene

To double check what a differentially expressed gene looks like, we can plot one with `DESeq2::plotCounts()` function.

```{r}
plotCounts(ddset, gene = "ENSMUSG00000026623", intgroup = "mutation_status")
```

The `R98S` mutated samples have higher expression of this gene than the control group, which helps assure us that the results are showing us what we are looking for. 

## Save results to TSV 

Write the results table to file.

```{r}
readr::write_tsv(
  deseq_df,
  file.path(
    results_dir,
    "SRP123625_differential_expression_results.tsv" # Replace with a relevant output file name
  )
)
```

## Create a volcano plot

We'll use the `EnhancedVolcano` package's main function to plot our data [@Zhu2018].

Here we are plotting the `log2FoldChange` (which was estimated by `lfcShrink` step) on the x axis and `padj` on the y axis.
The `padj` variable are the p values corrected with `Benjamini-Hochberg` (the default from the `results()` step).

Because we are using adjusted p values we can feel safe in making our `pCutoff` argument `0.01` (default is `1e-05`).  
Take a look at all the options for tailoring this plot using `?EnhancedVolcano`.

We will save the plot to our environment as `volcano_plot` to make it easier to save the figure separately later.

```{r}
# We'll assign this as `volcano_plot`
volcano_plot <- EnhancedVolcano::EnhancedVolcano(
  deseq_df,
  lab = deseq_df$Gene,
  x = "log2FoldChange",
  y = "padj",
  pCutoff = 0.01 # Loosen the cutoff since we supplied corrected p-values
)

# Print out plot here
volcano_plot
```

This looks pretty good! 
Let's save it to a PNG. 

```{r}
ggsave(
  plot = volcano_plot,
  file.path(plots_dir, "SRP123625_volcano_plot.png")
) # Replace with a plot name relevant to your data
```

Heatmaps are also a pretty common way to show differential expression results. 
You can take your results from this example and make a heatmap [following our heatmap module](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/clustering_rnaseq_01_heatmap.html).

# Further learning resources about this analysis

- [`DESeq2` vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
- [DESeq2 paper](https://doi.org/10.1186/s13059-014-0550-8) [@Love2014]
- [StatQuest Video: DESeq2, part 1, Library Normalization](https://www.youtube.com/watch?v=UFB993xufUU)
- [The `EnhancedVolcano` vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html) has more examples on how to tailor your volcano plot [@Blighe2020].

# Session info

At the end of every analysis, before saving your notebook, we recommend printing out your session info. 
This helps make your code more reproducible by recording what versions of software and packages you used to run this. 

```{r}
# Print session info
sessioninfo::session_info()
```

# References
