---
title: "PCA Visualization - RNA-seq"
author: "CCDL for ALSF"
date: "`r format(Sys.time(), '%B %Y')`"
output:   
  html_notebook:
    toc: true
    toc_float: true
    number_sections: true
---

# Purpose of the analysis



This notebook illustrates one way that you can use RNA-seq data from refine.bio to perform Principal Component Analysis (PCA) and plot the scores using `ggplot2`. 

# How to run this example

For general information about our tutorials and the basic software packages you will need, please see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-this-tutorial-is-structured).
We recommend taking a look at our [Resources for Learning R](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#resources-for-learning-r) if you have not written code in R before. 

## Obtain the `.Rmd` file of this analysis

To run this example yourself, [download the `.Rmd` for this analysis by clicking this link](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/dimension-reduction_rnaseq_01_pca.Rmd).

You can open this `.Rmd` file in RStudio and follow the rest of these steps from there. (See our [section about getting started with R notebooks](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) if you are unfamiliar with `.Rmd` files.)
Clicking this link will most likely send this to your downloads folder on your computer. 
Move this `.Rmd` file to where you would like this example and its files to be stored.

## Set up your analysis folders

Good file organization is helpful for keeping your data analysis project on track!
We have set up some code that will automatically set up a folder structure for you. 
Run this next chunk to set up your folders! 

If you have trouble running this chunk, see our [introduction to using `.Rmd`s](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-and-use-rmds) for more resources and explanations. 

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP133573") # Replace with path to desired data directory

# Create the data folder if it doesn't exist
if (!dir.exists(data_dir)) {
  dir.create(data_dir)
}

# Define the file path to the plots directory
plots_dir <- "plots" # Can replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results" # Can replace with path to desired output results directory

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

## Obtain the dataset from refine.bio

For general information about downloading data for these examples, see our ['Getting Started' section](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#how-to-get-the-data).

Go to this [dataset's page on refine.bio](https://www.refine.bio/experiments/SRP133573).

Click the "Download Now" button on the right side of this screen.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/dcce9a49246a15bfc7bb94a50a082c300c78074f/template/screenshots/download-now.png" width=200>  

Fill out the pop up window with your email and our Terms and Conditions:  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/dcce9a49246a15bfc7bb94a50a082c300c78074f/template/screenshots/download-email.png" width=500>  



We are going to use non-quantile normalized data for this analysis.
To get this data, you will need to check the box that says "Skip quantile normalization for RNA-seq samples".
Note that this option will only be available for RNA-seq datasets.

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/dcce9a49246a15bfc7bb94a50a082c300c78074f/template/screenshots/skip-quantile-normalization.png" width=500>

It may take a few minutes for the dataset to process.
You will get an email when it is ready. 

## About the dataset we are using for this example

For this example analysis, we will use this [prostate cancer dataset](https://www.refine.bio/experiments/SRP133573).

<!--BRIEFLY SUMMARIZE THIS DATASET--> 

The data that we downloaded from refine.bio for this analysis has 175 RNA-seq samples obtained from 20 patients with prostate cancer.
Patients underwent androgen deprivation therapy (ADT) and RNA-seq samples include pre-ADT biopsies and post-ADT prostatectomy specimens.

## Place the dataset in your new `data/` folder

Refine.bio will send you a download button in the email when it is ready. 
Follow the prompt to download a zip file that has a name with a series of letters and numbers and ends in `.zip`. 
Double clicking should unzip this for you and create a folder of the same name.  

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/dcce9a49246a15bfc7bb94a50a082c300c78074f/template/screenshots/download-folder-structure.png" width=400> 

For more details on the contents of this folder see [these docs on refine.bio](http://docs.refine.bio/en/latest/main_text.html#rna-seq-sample-compendium-download-folder).  

The `<experiment_accession_id>` folder has the data and metadata TSV files you will need for this example analysis.
Experiment accession ids usually look something like `GSE1235` or `SRP12345`. 

Copy and paste the `SRP133573` folder into your newly created `data/` folder.

## Check out our file structure!

Your new analysis folder should contain: 

- The example analysis Rmd you downloaded  
- A folder called "data" which contains:
  - The `SRP133573` folder which contains:
    - The gene expression  
    - The metadata TSV  
- A folder for `plots` (currently empty)  
- A folder for `results` (currently empty)  
Your example analysis folder should now look something like this (except with respective experiment accession id and analysis notebook name you are using): 

<img src="https://github.com/AlexsLemonade/refinebio-examples/raw/dcce9a49246a15bfc7bb94a50a082c300c78074f/template/screenshots/analysis-folder-structure.png" width=400>

In order for our example here to run without a hitch, we need these files to be in these locations so we've constructed a test to check before we get started with the analysis. 
Run this chunk to double check that your files are in the right place. 

```{r}
# Check if the gene expression matrix file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "SRP133573.tsv"))

# Check if the metadata file is in the data directory stored as `data_dir`
file.exists(file.path(data_dir, "metadata_SRP133573.tsv"))
```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots to the `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences. 

***

# PCA Visualization - RNA-seq

## Install libraries

See our Getting Started page with [instructions for package installation](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#what-you-need-to-install) for a list of the other software you will need, as well as more tips and resources.

In this analysis, we will be using the R package [`DESeq2`](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) [@Love2014] for normalization and production of PCA values and the R package [`ggplot2`](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html) [@Prabhakaran2016] for plotting the PCA values.

```{r}
if (!("DESeq2" %in% installed.packages())) {
  # Install DESeq2
  BiocManager::install("DESeq2", update = FALSE)
}
```

Attach the `DESeq2` and `ggplot2` libraries:

```{r}
# Attach the `DESeq2` library
library(DESeq2)

# Attach the `ggplot2` library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Set the seed so our results are reproducible:
set.seed(12345)
```

## Import and set up data

Data downloaded from refine.bio include a metadata tab separated values (TSV) file and a data TSV file. 
This chunk of code will read the both TSV files and add them as data frames to your environment.

```{r}
# Read in metadata TSV file
metadata <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your metadata file
  "metadata_SRP133573.tsv" # Replace with the name of your metadata file
))

# Read in data TSV file
df <- readr::read_tsv(file.path(
  data_dir, # Replace with path to your data file
  "SRP133573.tsv" # Replace with the name of your data file
)) %>%
  # Here we are going to store the gene IDs as rownames so that we can have a numeric matrix to perform calculations on later
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>% dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(df), metadata$refinebio_accession_code)
```

<!--ALL THE ANALYSIS STEPS HERE -->

Now we are going to use a combination of functions from the `DESeq2` and `ggplot2` packages to perform and visualize the results of the Principal Component Analysis (PCA) dimension reduction technique on our pre-ADT and post-ADT samples.

### Prepare data for `DESeq2`

We need to make sure all of the values in our data are converted to integers as required by a `DESeq2` function we will use later.

```{r}
# The `DESeqDataSetFromMatrix()` function needs the values to be converted to integers
df <- df %>%
  # Mutate numeric variables to be integers
  dplyr::mutate_if(is.numeric, round)
```

### Prepare metadata for `DESEq2`

We need to make sure all of the metadata column variables, that we would like to use to annotate our plot, are converted into factors.

```{r}
# We need to also format the variables from the metadata, that we will be using for annotation of the PCA plot, into factors
metadata <- metadata %>%
  dplyr::mutate(
    refinebio_treatment = as.factor(refinebio_treatment),
    refinebio_disease = as.factor(refinebio_disease)
  )
```

## Create a DESeqDataset

We will be using the `DESeq2` package for normalization, which requires us to format our data into a `DESeqDataSet` object.
We turn the data frame (or matrix) into a [`DESeqDataSet` object](TODO: Add link to header section) and specify which variable labels our experimental groups using the [`design` argument](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) [@Love2014].
In this chunk of code, we will not provide a specific model to the `design` argument because we are not performing a differential expression analysis.

```{r}
# Create a `DESeqDataSet` object
dds <- DESeqDataSetFromMatrix(
  countData = df, # This is the data frame with the counts values for all replicates in our dataset
  colData = metadata, # This is the data frame with the annotation data for the replicates in the counts data frame
  design = ~1 # Here we are not specifying a model -- Replace with an appropriate design variable for your analysis
)
```

## Define a minimum counts cutoff

We want to filter out the genes that have not been expressed or that have low expression counts because we want to remove any possible noise from our data before we normalize the data and create our plot.
We are going to do some pre-filtering to keep only genes with 10 or more reads total.
Note that rows represent gene data and the columns represent sample data in our dataset.

```{r}
# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
# only rows that have counts above the cutoff
genes_to_keep <- rowSums(counts(dds)) >= 10
dds <- dds[genes_to_keep, ]
```

## Perform DESeq2 normalization

We are going to use the `vst()` function from the `DESeq2` package to normalize the data.
TODO: Add a note/link to section in documentation on why we use `vst()` vs `rlog()` and `normTransform()`.

```{r}
# Normalize the data in the `DESeqDataSet` object using the `vst()` function from the `DESEq2` R package
dds_norm <- vst(dds)
```


## Create PCA plot using DESeq2

In this code chunk, the variable `refinebio_treatment` is given to the `plotPCA()` function as part of the goal of the experiment is to analyze the sample transcriptional responses to androgen deprivation therapy (ADT).

```{r}
plotPCA(dds_norm,
  intgroup = "refinebio_treatment"
)
```

In this chunk, we are going to add another variable to our plot for annotation.

Now we'll plot the PCA using both `refinebio_treatment` and `refinebio_disease` variables for labels since they are central to the androgen deprivation therapy (ADT) based hypothesis in the [original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6210624/) [@Sharma2018].

```{r}
plotPCA(dds_norm,
  intgroup = c("refinebio_treatment", "refinebio_disease") # Note that we are able to add another variable to the intgroup argument here by providing a vector of the variable names with  `c()` function
)
```

In the plot above, it is hard to distinguish the different `refinebio_treatment` values which contain the data on whether or not samples have been treated with ADT versus the `refinebio_disease` values which refer to the method by which the samples were obtained from patients (i.e. biopsy).

Let's use the `ggplot2` package functionality to customize our plot further and make the annotation labels better distinguishable.

First let's use `plotPCA()` to receive and store the PCA values for plotting.

```{r}
# We first have to save the results of the `plotPCA()` function for use with `ggplot2`
pcaData <-
  plotPCA(
    dds_norm,
    intgroup = c("refinebio_treatment", "refinebio_disease"),
    returnData = TRUE # This argument tells R to return the PCA values
  )
```

Now let's plot our `pcaData` using `ggplot2` functionality.

```{r}
# Plot using `ggplot()` function
annotated_pca_plot <- ggplot(
  pcaData,
  aes(
    x = PC1,
    y = PC2,
    color = refinebio_treatment, # This will label points with different colors for each `refinebio_disease` group
    shape = refinebio_disease # This will label points with different shapes for each `refinebio_disease` group
  )
)
```

## Save annotated PCA plot as a PNG

You can easily switch this to save to a JPEG or TIFF by changing the file name within the `ggsave()` function to the respective file suffix.

```{r}
# Save plot using `ggsave()` function
ggsave(file.path(plots_dir, "SRP133573_pca_plot.png"), # Replace with name relevant your plotted data
  plot = annotated_pca_plot # Here we are giving the function the plot object that we want saved to file
)
```

# Resources for further learning

- [Principle Component Analysis (PCA) Explained Visually](http://setosa.io/ev/principal-component-analysis/) [@pca-visually-explained]
- [Guidelines on choosing dimension reduction methods](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1006907&type=printable) [@Nguyen2019]  
- [A nice explanation and comparison of many different dimenstionality reduction techniques that you may encounter](https://rpubs.com/Saskia/520216) [@Freytag2019]

# Print session info

At the end of every analysis, before saving your notebook, we recommend printing out your session info. 
This helps make your code more reproducible by recording what versions of softwares and packages you used to run this. 

```{r}
# Print session info
sessionInfo()
```
